{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Computational biology and bioinformatics - <span style=\"color:#1CA766\">INFO-F-439</span>\n",
    "# Assignment 1: <span style=\"color:#1CA766\">GOR III secondary structure prediction</span>\n",
    "\n",
    "> ## <span style=\"color:#2E66A7\"> Alberto Parravicini</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*****\n",
    "\n",
    "# <span style=\"color:#2E66A7\">Part 1:</span> Introduction\n",
    "\n",
    "Proteins are for the most part defined by the aminoacid sequences that compose them. However, due to the nature of the aminoacids, proteins have more complex structures that characterize them.\n",
    "\n",
    "Hydrogen bonds between the aminoacids in the sequence causes the proteins to assume three-dimensional structures of various kind. Being able to recognize these structures is useful to compare proteins, and to get a better insight on their properties.\n",
    "\n",
    "Classifying correctly which portion of a protein assumes a given secondary structure isn't an easy task, but it can be seen that there are links between specific patterns of aminoacids and given secondary structures.\n",
    "As an example, in $\\alpha$-helices (one type of secondary structure) every third aminoacid along the sequence will tend to be hydrophobic. Also, regions richer of *alanine (A)* and *leucine (L)* will tend to form helices. (see *Gollery, Martin. \"Bioinformatics: Sequence and Genome Analysis\", David W. Mount. Pag. 387-388*)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "### <span style=\"color:#1CA766\">GOR I & II</span>:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One idea that can be exploited to predict secondary structures from aminoacid sequeces, is that the secondary structure corresponding to a certain aminoacid is influenced not only by the aminoacid itself, but also by the aminoacids that are close to it.\n",
    "\n",
    "Given a dataset of proteins where for each aminoacid its secondary structure is known, it is possible to build a model that is able to predict secondary sequences, by using the idea above.\n",
    "\n",
    "This is the core idea behind the **GOR I** and **GOR II** algorithms.\n",
    "\n",
    "For a given position $j$, it will be predicted the secondary structure $S_j$ that has the highest value of *information*, as function of the aminoacid sequence.\n",
    "Assuming that our secondary structures are in the set $S = \\{\\alpha$-helices (**H**), $\\beta$-sheets (**B**), and coils (**C**)$\\}$, the one that will be chosen is $$S_j = \\arg\\max_{S}{I(\\Delta S_j;\\ a_1, \\ldots, a_n)}$$ where $a_1, \\ldots, a_n$ is the aminoacid sequence, and $S_j$ is picked among the secondary structures in S.\n",
    "\n",
    "The previous computation is approximated by **GOR** by considering a window of size $17$ around position $j$, which gives the formula $$S_j = \\arg\\max_{S}{I(\\Delta S_j;\\ a_1, \\ldots, a_n)} \\approx \\arg\\max_{S}{\\sum_{m = -8}^{m = 8}{I(\\Delta S_j;\\ a_{j+m})}}$$ \n",
    "\n",
    "In the previous formulation, $I(\\Delta S_j;\\ a_{j+m})$ is a measure of *self-information*, and is given by\n",
    "$$I(\\Delta S_j;\\ a_{j+m}) = log(\\frac{c_{S_j,\\ a_{j+m}}}{c_{\\neg S_j,\\ a_{j+m}}}) + log(\\frac{c_{\\neg S}}{c_{S}})$$\n",
    "\n",
    "where $c$ refers to the count of the occurrences (in a given dataset) of what is specified in the subscript; for instance, $c_{\\neg S}$ is the number of aminoacids with a secondary structure different from $S$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In Python, we can compute $c_{S_j,\\ a_{j+m}}$ and $I(\\Delta S_j;\\ a_{j+m})$ in the following way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def count_residues_gor_1(s_list, a_list, s_j, a_jm, lag):\n",
    "    \"\"\"\n",
    "    Count the number of times that structure \"s_j\" appears,\n",
    "    and residue \"a_jm\" is at distance \"lag\" from them.\n",
    "    :param s_list: a list of secondary structures; pandas Series.\n",
    "    :param a_list: a list of aminoacids; pandas Series.\n",
    "    :param s_j: a secondary structure name.\n",
    "    :param a_jm: an aminoacid name.\n",
    "    :param lag: distance from a_j at which a_jm is searched.\n",
    "    :return: int\n",
    "    \"\"\"\n",
    "    n = len(s_list)\n",
    "    count = 0\n",
    "    for i in range(max(0, -lag), min(n, n - lag)):\n",
    "        if s_list.iloc[i] == s_j and a_list.iloc[i + lag] == a_jm:\n",
    "            count += 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def info_value_gor_1(s_list, a_list, s_j, a_jm, lag, a_occ):\n",
    "    \"\"\"\n",
    "    Compute the information value associated to a secondary structure s_j, in position j,\n",
    "    and to an aminoacid a_jm, at distance m from s_j.\n",
    "    :param s_list: a list of secondary structures; pandas Series.\n",
    "    :param a_list: a list of aminoacids; pandas Series.\n",
    "    :param s_j: a secondary structure name.\n",
    "    :param a_jm: an aminoacid name.\n",
    "    :param lag: distance from a_j at which a_jm is searched.\n",
    "    :param a_occ: number of occurrencies of each aminoacid; pandas Series.\n",
    "    :return: double\n",
    "    \"\"\"\n",
    "    num_1 = count_residues_gor_1(s_list, a_list, s_j, a_jm, lag)\n",
    "    den_1 = a_occ[a_jm] - num_1\n",
    "    den_2 = s_occ[s_j]\n",
    "    num_2 = len(s_list) - den_2\n",
    "    return np.log(num_1 * num_2) - np.log(den_1 * den_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "### <span style=\"color:#1CA766\">GOR III</span>:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is possible to improve the accuracy of **GOR I & II** by considering not only the occurrencies of aminoacids close to the position where we want to predict the secondary structure, but also their occurrencies relatively to the aminoacid found at that given position.\n",
    "\n",
    "This means that the information value of a secondary structure $S_j$ will be approximated as \n",
    "\n",
    "$$S_j = \\arg\\max_{S}{I(\\Delta S_j;\\ a_1, \\ldots, a_n)} \\approx \\arg\\max_{S}{\\sum_{m = -8}^{m = 8}{I(\\Delta S_j;\\ a_{j+m}\\ |\\ a_j)}} = \\ldots$$ \n",
    "$$\\ldots = I(\\Delta S_j;\\ a_j) + \\sum_{m = -8,\\ m \\neq 0}^{m = 8}{I(\\Delta S_j;\\ a_{j+m}\\ |\\ a_j)}$$\n",
    "\n",
    "where it holds $$I(\\Delta S_j;\\ a_{j+m}\\ |\\ a_j) = log(\\frac{c_{S_j,\\ a_{j+m},\\ a_j}}{c_{\\neg S_j,\\ a_{j+m},\\ a_j}}) + log(\\frac{c_{\\neg S_j,\\ a_j}}{c_{S_j,\\ a_j}})$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In Python, we can compute $c_{S_j,\\ a_{j+m},\\ a_j}$ and $I(\\Delta S_j;\\ a_{j+m}\\ |\\ a_j)$ as:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def count_residues_gor_3(s_list, a_list, s_j, a_j, a_jm, lag):\n",
    "    \"\"\"\n",
    "    Count the number of times that structure \"s_j\" appears together with residue \"a_j\",\n",
    "    and residue \"a_jm\" is at distance \"lag\" from them.\n",
    "    :param s_list: a list of secondary structures; pandas Series.\n",
    "    :param a_list: a list of aminoacids; pandas Series.\n",
    "    :param s_j: a secondary structure name\n",
    "    :param a_j: an aminoacid name\n",
    "    :param a_jm: an aminoacid name\n",
    "    :param lag: distance from a_j at which a_jm is searched\n",
    "    :return: int\n",
    "    \"\"\"\n",
    "    n = len(s_list)\n",
    "    count = 0\n",
    "    for i in range(max(0, -lag), min(n, n - lag)):\n",
    "        if s_list.iat[i] == s_j and a_list.iat[i] == a_j and a_list.iat[i + lag] == a_jm:\n",
    "            count += 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def info_value_gor_3(s_list, a_list, s_j, a_j, a_jm, lag, a_occ):\n",
    "    \"\"\"\n",
    "    Compute the information value associated to an aminoacid a_j in position j,\n",
    "    with secondary structure s_j, in position j,\n",
    "    and to an aminoacid a_jm, at distance m from s_j.\n",
    "    :param s_list: a list of secondary structures; pandas Series.\n",
    "    :param a_list: a list of aminoacids; pandas Series.\n",
    "    :param s_j: a secondary structure name\n",
    "    :param a_j: an aminoacid name\n",
    "    :param a_jm: an aminoacid name\n",
    "    :param lag: distance from a_j at which a_jm is searched\n",
    "    :param a_occ: number of occurrencies of each aminoacid\n",
    "    :return: double\n",
    "    \"\"\"\n",
    "    # Num of times s_j appears with a_j, and a_jm is lag position distant.\n",
    "    num_1 = count_residues_gor_3(s_list, a_list, s_j, a_j, a_jm, lag)\n",
    "    # Num of times a_j appears with a secondary structure different from s_j,\n",
    "    # and a_jm is lag position distant.\n",
    "    den_1 = count_aminoacid_lag(a_list, a_j, a_jm, lag) - num_1\n",
    "    \n",
    "    # Number of times aminoacid a_j appears together with structure s_j.                        \n",
    "    den_2 = count_residues_gor_3(s_list, a_list, s_j, a_j, a_j, 0)                          \n",
    "    # Number of times aminoacid a_j appears with a different structure from s_j.\n",
    "    num_2 = a_occ[a_j] - den_2\n",
    "                 \n",
    "    info = np.log(num_1) - np.log(den_1) + np.log(num_2) - np.log(den_2)   \n",
    "    return info     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "*****\n",
    "\n",
    "# <span style=\"color:#2E66A7\">Part 2:</span> Data pre-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To compute the information values for each combination of aminoacids and secondary structures, it is necessary to have a large dataset of proteins with known secondary structures. We are provided with a set of 498 proteins, where the secondary structures have been determined by **DSSP** and **STRIDE**.\n",
    "\n",
    "Before being able to apply **GOR III**, it is required to preprocess the dataset, clean the inconsistencies that might be present, and store the aminoacid sequences in an appropriate way.\n",
    "\n",
    "It can be noted that there are a few rows with aminoacids that are not part of the usual 20. As the number of these rows is very limited compared to the total, we can just remove them entirely, and this won't affect the overall results of the algorithm."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#%% IMPORT DATA\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import timeit\n",
    "import pickle\n",
    "\n",
    "aminoacid_list = \"arndceqghilkmfpstwyv\"\n",
    "\n",
    "# Dictionary that maps aminoacid codes to single letters.\n",
    "aminoacid_codes = {\"ala\":  \"a\",\n",
    "                   \"arg\":  \"r\",\n",
    "                   \"asn\":  \"n\",\n",
    "                   \"asp\":  \"d\",\n",
    "                   \"cys\":  \"c\",\n",
    "                   \"gln\":  \"q\",\n",
    "                   \"glu\":  \"e\",\n",
    "                   \"gly\":  \"g\",\n",
    "                   \"his\":  \"h\",\n",
    "                   \"ile\":  \"i\",\n",
    "                   \"leu\":  \"l\",\n",
    "                   \"lys\":  \"k\",\n",
    "                   \"met\":  \"m\",\n",
    "                   \"phe\":  \"f\",\n",
    "                   \"pro\":  \"p\",\n",
    "                   \"ser\":  \"s\",\n",
    "                   \"thr\":  \"t\",\n",
    "                   \"trp\":  \"w\",\n",
    "                   \"tyr\":  \"y\",\n",
    "                   \"val\":  \"v\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def preprocess_input(file_name, aminoacid_codes):\n",
    "    \"\"\"\n",
    "    Load and preprocess a file containing protein sequences.\n",
    "    The output will be a DataFrame ready to be used by GOR.\n",
    "    :param file_name: string, name of the file to be opened.\n",
    "    :param aminoacid_codes: dictionary that maps aminoacid codes to single letters.\n",
    "    :return: DataFrame\n",
    "    \"\"\"\n",
    "    # Load the data\n",
    "    input_data = pd.read_csv(file_name,\n",
    "                         header=None, sep=\"\\t\",\n",
    "                         names=[\"PDB_code\", \"PDB_chain_code\", \"PDB_seq_code\", \"residue_name\", \"secondary_structure\"])\n",
    "\n",
    "    # CLEAN DATA - AMINOACIDS\n",
    "\n",
    "    # There are some values that are weird, like \"a\", \"b\", ...\n",
    "    # Some of those can be interpreted as correct aminoacids.\n",
    "\n",
    "    # Remove rows with \"X\", \"b\", \"UNK\"\n",
    "    # First, remove any leading spaces, then remove \"X\", \"b\", \"UNK\"\n",
    "    input_data[\"residue_name\"] = input_data[\"residue_name\"].str.strip()\n",
    "    input_data = input_data.drop(input_data[input_data[\"residue_name\"].isin([\"X\", \"b\", \"UNK\"])].index).reset_index(drop=True)\n",
    "\n",
    "\n",
    "    # Add a column for the aminoacids, while preserving the original one\n",
    "    input_data[\"a\"] = input_data[\"residue_name\"]\n",
    "    input_data[\"a\"] = input_data[\"a\"].str.lower()\n",
    "\n",
    "\n",
    "    # replace the codes with something shorter\n",
    "    input_data[\"a\"] = input_data[\"a\"].map(lambda x: aminoacid_codes[x] if x in aminoacid_codes else x)\n",
    "\n",
    "    # Create a new column\n",
    "    input_data[\"s\"] = input_data[\"secondary_structure\"]\n",
    "    input_data[\"s\"] = input_data[\"s\"].str.lower()\n",
    "    # Replace \"other\" with \"coil\"\n",
    "    input_data.loc[input_data[\"s\"] == \"other\", \"s\"] = \"coil\"\n",
    "    # Shorten values\n",
    "    input_data[\"s\"] = input_data[\"s\"].str[0]\n",
    "    \n",
    "    # Add a composite key to the dataset.\n",
    "    # Append the PDB chain code to the PDB code, to obtain unique protein identifier.\n",
    "    input_data[\"PDB_code_and_chain\"] = input_data.PDB_code + \"_\" + input_data.PDB_chain_code\n",
    "\n",
    "    return input_data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can load and work on the *STRIDE* dataset. The *DSSP* dataset is similar, and will be compared to *STRIDE* later in the report. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CYS', 'LYS', 'VAL', 'TYR', 'MET', 'PHE', 'ARG', 'SER', 'GLY', 'GLN', 'ASN', 'GLU', 'ALA', 'THR', 'PRO', 'HIS', 'ILE', 'ASP', 'LEU', 'TRP'}\n",
      "{'Other', 'Beta', 'Helix'}\n"
     ]
    }
   ],
   "source": [
    "    # Type of the data to read (\"stride\", \"dssp\")\n",
    "    data_type = \"stride\"\n",
    "    # File name\n",
    "    file_name = \"../data/\" + data_type + \"_info.txt\"\n",
    "    # Read the data\n",
    "    input_data = preprocess_input(file_name, aminoacid_codes)\n",
    "\n",
    "    # Look at the aminoacid values\n",
    "    aminoacids = set(input_data.residue_name)\n",
    "    print(aminoacids)\n",
    "    # Same stuff with the secondary structures\n",
    "    secondary_structures = set(input_data.secondary_structure)\n",
    "    print(secondary_structures)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PDB_code</th>\n",
       "      <th>PDB_chain_code</th>\n",
       "      <th>PDB_seq_code</th>\n",
       "      <th>residue_name</th>\n",
       "      <th>secondary_structure</th>\n",
       "      <th>a</th>\n",
       "      <th>s</th>\n",
       "      <th>PDB_code_and_chain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1w0n</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>ILE</td>\n",
       "      <td>Other</td>\n",
       "      <td>i</td>\n",
       "      <td>c</td>\n",
       "      <td>1w0n_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1w0n</td>\n",
       "      <td>A</td>\n",
       "      <td>13</td>\n",
       "      <td>THR</td>\n",
       "      <td>Beta</td>\n",
       "      <td>t</td>\n",
       "      <td>b</td>\n",
       "      <td>1w0n_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1w0n</td>\n",
       "      <td>A</td>\n",
       "      <td>14</td>\n",
       "      <td>LYS</td>\n",
       "      <td>Beta</td>\n",
       "      <td>k</td>\n",
       "      <td>b</td>\n",
       "      <td>1w0n_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1w0n</td>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>VAL</td>\n",
       "      <td>Beta</td>\n",
       "      <td>v</td>\n",
       "      <td>b</td>\n",
       "      <td>1w0n_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1w0n</td>\n",
       "      <td>A</td>\n",
       "      <td>16</td>\n",
       "      <td>GLU</td>\n",
       "      <td>Beta</td>\n",
       "      <td>e</td>\n",
       "      <td>b</td>\n",
       "      <td>1w0n_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PDB_code PDB_chain_code  PDB_seq_code residue_name secondary_structure  a  \\\n",
       "0     1w0n              A            12          ILE               Other  i   \n",
       "1     1w0n              A            13          THR                Beta  t   \n",
       "2     1w0n              A            14          LYS                Beta  k   \n",
       "3     1w0n              A            15          VAL                Beta  v   \n",
       "4     1w0n              A            16          GLU                Beta  e   \n",
       "\n",
       "   s PDB_code_and_chain  \n",
       "0  c             1w0n_A  \n",
       "1  b             1w0n_A  \n",
       "2  b             1w0n_A  \n",
       "3  b             1w0n_A  \n",
       "4  b             1w0n_A  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_data.iloc[0:5, ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**PDB_code** and **PDB_chain_code** work as identifier of the row. The column **a** and **s** are just shortened versions of the aminoacid codes and of the secondary structure codes.\n",
    "\n",
    "We can also compute some other values, that will be useful later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Number of residues\n",
    "n_res = len(input_data.index)\n",
    "\n",
    "# Compute the number of occurrencies of each aminoacid\n",
    "a_occ = input_data[\"a\"].value_counts()\n",
    "\n",
    "# Compute the number of occurrencies of each secondary structure\n",
    "s_occ = input_data[\"s\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "71073"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "l    6449\n",
       "a    5963\n",
       "g    5119\n",
       "v    4960\n",
       "e    4816\n",
       "d    4244\n",
       "k    4213\n",
       "s    4148\n",
       "i    4034\n",
       "t    3937\n",
       "r    3665\n",
       "p    3181\n",
       "n    3053\n",
       "f    2890\n",
       "q    2849\n",
       "y    2467\n",
       "h    1715\n",
       "m    1260\n",
       "c    1103\n",
       "w    1007\n",
       "Name: a, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a_occ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "c    29170\n",
       "h    25755\n",
       "b    16148\n",
       "Name: s, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_occ"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, to compute the information values of **GOR III** it is required to add a few more functions.\n",
    "\n",
    "The first will compute a matrix that stores, for each aminoacid, the occurrences of each secondary structure.  \n",
    "The second will build a dictionary where each key corresponds to a secondary structure, and each value is a tensor (technically, a pandas.Panel) that stores the number of times where an aminoacid $a_j$ appears with secondary structure $S_j$,\n",
    "and another aminoacid $a_{j+m}$ appears at a distance $m$ from them, with $m$ in $[-8, 8]$.\n",
    "\n",
    "It is necessary to precompute these matrices as counting from scratch the occurrencies, for each prediction, could be very inefficient."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def build_sj_aj_matrix(s_list, a_list):\n",
    "    \"\"\"\n",
    "    Count the number of times that each aminoacid a_j has secondary structure s_j.\n",
    "    :param s_list: a list of secondary structures; pandas Series.\n",
    "    :param a_list: a list of aminoacids; pandas Series.\n",
    "    :return: pandas.DataFrame\n",
    "    \"\"\"\n",
    "    # Set of all secondary structures.\n",
    "    sec_structure_set = set(s_list)\n",
    "    # Set of all aminoacids.\n",
    "    aminoacid_set = set(a_list)\n",
    "    # Occurrency matrix.\n",
    "    sj_aj_matrix = pd.DataFrame(0, index=sec_structure_set, columns=aminoacid_set, dtype=int)\n",
    "    # Zip together the lists of secondary structures and aminoacids, it makes counting easier.\n",
    "    s_a_list = [list(x) for x in zip(s_list,a_list)]\n",
    "    # Count the occurrencies.\n",
    "    for s in sec_structure_set:\n",
    "        for a in aminoacid_set:\n",
    "            sj_aj_matrix.at[s, a] = s_a_list.count([s, a])\n",
    "\n",
    "    return sj_aj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "! -> EXECUTION TIME OF build_sj_aj_matrix: 0.26465835296067575 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time = timeit.default_timer()\n",
    "sj_aj_matrix = build_sj_aj_matrix(input_data.s, input_data.a)\n",
    "end_time = timeit.default_timer()\n",
    "print(\"! -> EXECUTION TIME OF build_sj_aj_matrix:\", (end_time - start_time), \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>i</th>\n",
       "      <th>v</th>\n",
       "      <th>n</th>\n",
       "      <th>l</th>\n",
       "      <th>f</th>\n",
       "      <th>c</th>\n",
       "      <th>t</th>\n",
       "      <th>p</th>\n",
       "      <th>a</th>\n",
       "      <th>s</th>\n",
       "      <th>g</th>\n",
       "      <th>e</th>\n",
       "      <th>k</th>\n",
       "      <th>h</th>\n",
       "      <th>q</th>\n",
       "      <th>y</th>\n",
       "      <th>r</th>\n",
       "      <th>d</th>\n",
       "      <th>w</th>\n",
       "      <th>m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>1572</td>\n",
       "      <td>2115</td>\n",
       "      <td>447</td>\n",
       "      <td>1573</td>\n",
       "      <td>938</td>\n",
       "      <td>301</td>\n",
       "      <td>1135</td>\n",
       "      <td>335</td>\n",
       "      <td>1007</td>\n",
       "      <td>771</td>\n",
       "      <td>804</td>\n",
       "      <td>773</td>\n",
       "      <td>809</td>\n",
       "      <td>400</td>\n",
       "      <td>493</td>\n",
       "      <td>827</td>\n",
       "      <td>742</td>\n",
       "      <td>530</td>\n",
       "      <td>294</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h</th>\n",
       "      <td>1468</td>\n",
       "      <td>1601</td>\n",
       "      <td>916</td>\n",
       "      <td>3071</td>\n",
       "      <td>1012</td>\n",
       "      <td>305</td>\n",
       "      <td>1089</td>\n",
       "      <td>467</td>\n",
       "      <td>2992</td>\n",
       "      <td>1204</td>\n",
       "      <td>850</td>\n",
       "      <td>2348</td>\n",
       "      <td>1795</td>\n",
       "      <td>570</td>\n",
       "      <td>1369</td>\n",
       "      <td>850</td>\n",
       "      <td>1586</td>\n",
       "      <td>1338</td>\n",
       "      <td>377</td>\n",
       "      <td>547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>994</td>\n",
       "      <td>1244</td>\n",
       "      <td>1690</td>\n",
       "      <td>1805</td>\n",
       "      <td>940</td>\n",
       "      <td>497</td>\n",
       "      <td>1713</td>\n",
       "      <td>2379</td>\n",
       "      <td>1964</td>\n",
       "      <td>2173</td>\n",
       "      <td>3465</td>\n",
       "      <td>1695</td>\n",
       "      <td>1609</td>\n",
       "      <td>745</td>\n",
       "      <td>987</td>\n",
       "      <td>790</td>\n",
       "      <td>1337</td>\n",
       "      <td>2376</td>\n",
       "      <td>336</td>\n",
       "      <td>431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      i     v     n     l     f    c     t     p     a     s     g     e  \\\n",
       "b  1572  2115   447  1573   938  301  1135   335  1007   771   804   773   \n",
       "h  1468  1601   916  3071  1012  305  1089   467  2992  1204   850  2348   \n",
       "c   994  1244  1690  1805   940  497  1713  2379  1964  2173  3465  1695   \n",
       "\n",
       "      k    h     q    y     r     d    w    m  \n",
       "b   809  400   493  827   742   530  294  282  \n",
       "h  1795  570  1369  850  1586  1338  377  547  \n",
       "c  1609  745   987  790  1337  2376  336  431  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj_aj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def build_sj_aj_ajm(input_data, print_details=True):\n",
    "    \"\"\"\n",
    "    Count the times where aminoacid a_j appears with secondary structure s_j,\n",
    "    and aminoacid a_jm appears at a distance m from them, with m in [-8, 8].\n",
    "    \"\"\"\n",
    "    # Set of all proteins.\n",
    "    protein_set = list(set(input_data.PDB_code_and_chain))\n",
    "    # Set of all secondary structures.\n",
    "    sec_structure_set = list(set(input_data.s))\n",
    "    # Set of all aminoacids.\n",
    "    aminoacid_set = list(set(input_data.a))\n",
    "    # Build a dictionary where the keys are the secondary structures, and the values are 3D tensors\n",
    "    # with index a_j, a_jm, m.\n",
    "    sj_aj_ajm_dict = {s: pd.Panel(data=0, items=aminoacid_set, major_axis=aminoacid_set, minor_axis=np.arange(-8, 9), dtype=int) for s in sec_structure_set}\n",
    "\n",
    "    # The counting must be done separately for each protein,\n",
    "    # otherwise the aminoacids at the end of one protein would be counted\n",
    "    # as part of the next one!\n",
    "    for i_p, p in enumerate(protein_set):\n",
    "        s_list = input_data.loc[input_data['PDB_code_and_chain'] == p].s\n",
    "        a_list = input_data.loc[input_data['PDB_code_and_chain'] == p].a\n",
    "        n = len(s_list)\n",
    "        for i_aj, a_j in enumerate(aminoacid_set):\n",
    "            for i_ajm, a_jm in enumerate(aminoacid_set):\n",
    "                for i_m, m in enumerate(np.arange(-8, 9)):\n",
    "                    if print_details:\n",
    "                        print(p, a_j, a_jm, m, \"-------\", i_p / len(protein_set), i_aj / 20, i_ajm / 20)\n",
    "\n",
    "                    for i in range(max(0, -m), min(n, n - m)):\n",
    "                        if a_list.iat[i] == a_j and a_list.iat[i + m] == a_jm:\n",
    "                            # Increment the count of the right dictionary.\n",
    "                            sj_aj_ajm_dict[s_list.iat[i]].iat[i_aj, i_ajm, i_m] += 1\n",
    "\n",
    "    return sj_aj_ajm_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "! -> EXECUTION TIME OF build_sj_aj_ajm: 5.784613011439177e-05 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Count the times where aminoacid a_j appears with secondary structure s_j,\n",
    "# and aminoacid a_jm appears at a distance m from them, with m in [-8, 8].\n",
    "#   WARNING: this can take a bit of time, usually about 20-30 minutes;\n",
    "#            do it at your own risk!\n",
    "start_time = timeit.default_timer()\n",
    "##########################\n",
    "# UNCOMMENT IF NEEDED ####\n",
    "##########################\n",
    "# sj_aj_ajm_dict = build_sj_aj_ajm(input_data)\n",
    "\n",
    "end_time = timeit.default_timer()\n",
    "print(\"! -> EXECUTION TIME OF build_sj_aj_ajm:\", (end_time - start_time), \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As the computation of the dictionary can take a fair bit of time, it is a good idea to store the result, and load it when needed.\n",
    "\n",
    "Note: comment/uncomment the following lines depending on your needs!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Save data\n",
    "dict_file_name = \"sj_aj_ajm_dict_\"+ data_type + \".p\"\n",
    "\n",
    "# # Open the file for writing.\n",
    "# file_object = open(dict_file_name.encode('utf-8').strip(), 'wb')\n",
    "\n",
    "# # Save data\n",
    "# pickle.dump(sj_aj_ajm_dict, file_object)\n",
    "#\n",
    "# file_object.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data can also be loaded with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Load data\n",
    "file_object = open(dict_file_name.encode('utf-8').strip(), 'rb')\n",
    "sj_aj_ajm_dict = pickle.load(file_object)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we can visualize a slice of **sj_aj_ajm_dict**: for instance, we can consider the occurrences of a $\\beta$-sheets with an aminoacid distance of $-7$. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g</th>\n",
       "      <th>d</th>\n",
       "      <th>f</th>\n",
       "      <th>w</th>\n",
       "      <th>r</th>\n",
       "      <th>q</th>\n",
       "      <th>c</th>\n",
       "      <th>n</th>\n",
       "      <th>a</th>\n",
       "      <th>e</th>\n",
       "      <th>p</th>\n",
       "      <th>t</th>\n",
       "      <th>v</th>\n",
       "      <th>y</th>\n",
       "      <th>l</th>\n",
       "      <th>i</th>\n",
       "      <th>k</th>\n",
       "      <th>s</th>\n",
       "      <th>m</th>\n",
       "      <th>h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>g</th>\n",
       "      <td>67</td>\n",
       "      <td>51</td>\n",
       "      <td>69</td>\n",
       "      <td>25</td>\n",
       "      <td>42</td>\n",
       "      <td>37</td>\n",
       "      <td>23</td>\n",
       "      <td>24</td>\n",
       "      <td>79</td>\n",
       "      <td>58</td>\n",
       "      <td>30</td>\n",
       "      <td>77</td>\n",
       "      <td>156</td>\n",
       "      <td>59</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "      <td>64</td>\n",
       "      <td>63</td>\n",
       "      <td>29</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>43</td>\n",
       "      <td>41</td>\n",
       "      <td>62</td>\n",
       "      <td>18</td>\n",
       "      <td>43</td>\n",
       "      <td>24</td>\n",
       "      <td>14</td>\n",
       "      <td>24</td>\n",
       "      <td>62</td>\n",
       "      <td>39</td>\n",
       "      <td>22</td>\n",
       "      <td>56</td>\n",
       "      <td>139</td>\n",
       "      <td>47</td>\n",
       "      <td>84</td>\n",
       "      <td>88</td>\n",
       "      <td>38</td>\n",
       "      <td>42</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>43</td>\n",
       "      <td>14</td>\n",
       "      <td>31</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "      <td>19</td>\n",
       "      <td>33</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>37</td>\n",
       "      <td>65</td>\n",
       "      <td>33</td>\n",
       "      <td>56</td>\n",
       "      <td>53</td>\n",
       "      <td>33</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>w</th>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>18</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>21</td>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r</th>\n",
       "      <td>30</td>\n",
       "      <td>28</td>\n",
       "      <td>31</td>\n",
       "      <td>17</td>\n",
       "      <td>45</td>\n",
       "      <td>22</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>40</td>\n",
       "      <td>56</td>\n",
       "      <td>16</td>\n",
       "      <td>64</td>\n",
       "      <td>106</td>\n",
       "      <td>47</td>\n",
       "      <td>91</td>\n",
       "      <td>64</td>\n",
       "      <td>35</td>\n",
       "      <td>35</td>\n",
       "      <td>8</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>28</td>\n",
       "      <td>16</td>\n",
       "      <td>25</td>\n",
       "      <td>9</td>\n",
       "      <td>32</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>20</td>\n",
       "      <td>36</td>\n",
       "      <td>27</td>\n",
       "      <td>8</td>\n",
       "      <td>51</td>\n",
       "      <td>81</td>\n",
       "      <td>33</td>\n",
       "      <td>60</td>\n",
       "      <td>56</td>\n",
       "      <td>23</td>\n",
       "      <td>30</td>\n",
       "      <td>6</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>15</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>41</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n</th>\n",
       "      <td>38</td>\n",
       "      <td>20</td>\n",
       "      <td>35</td>\n",
       "      <td>18</td>\n",
       "      <td>42</td>\n",
       "      <td>30</td>\n",
       "      <td>14</td>\n",
       "      <td>22</td>\n",
       "      <td>45</td>\n",
       "      <td>28</td>\n",
       "      <td>11</td>\n",
       "      <td>56</td>\n",
       "      <td>74</td>\n",
       "      <td>35</td>\n",
       "      <td>65</td>\n",
       "      <td>78</td>\n",
       "      <td>40</td>\n",
       "      <td>24</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>63</td>\n",
       "      <td>29</td>\n",
       "      <td>62</td>\n",
       "      <td>20</td>\n",
       "      <td>44</td>\n",
       "      <td>29</td>\n",
       "      <td>17</td>\n",
       "      <td>37</td>\n",
       "      <td>95</td>\n",
       "      <td>58</td>\n",
       "      <td>22</td>\n",
       "      <td>89</td>\n",
       "      <td>178</td>\n",
       "      <td>59</td>\n",
       "      <td>131</td>\n",
       "      <td>133</td>\n",
       "      <td>48</td>\n",
       "      <td>54</td>\n",
       "      <td>27</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>e</th>\n",
       "      <td>53</td>\n",
       "      <td>30</td>\n",
       "      <td>59</td>\n",
       "      <td>19</td>\n",
       "      <td>45</td>\n",
       "      <td>36</td>\n",
       "      <td>19</td>\n",
       "      <td>32</td>\n",
       "      <td>63</td>\n",
       "      <td>43</td>\n",
       "      <td>21</td>\n",
       "      <td>64</td>\n",
       "      <td>147</td>\n",
       "      <td>42</td>\n",
       "      <td>104</td>\n",
       "      <td>101</td>\n",
       "      <td>49</td>\n",
       "      <td>46</td>\n",
       "      <td>15</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>40</td>\n",
       "      <td>15</td>\n",
       "      <td>43</td>\n",
       "      <td>13</td>\n",
       "      <td>36</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "      <td>19</td>\n",
       "      <td>49</td>\n",
       "      <td>34</td>\n",
       "      <td>19</td>\n",
       "      <td>45</td>\n",
       "      <td>81</td>\n",
       "      <td>38</td>\n",
       "      <td>65</td>\n",
       "      <td>50</td>\n",
       "      <td>37</td>\n",
       "      <td>40</td>\n",
       "      <td>12</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t</th>\n",
       "      <td>47</td>\n",
       "      <td>28</td>\n",
       "      <td>44</td>\n",
       "      <td>13</td>\n",
       "      <td>41</td>\n",
       "      <td>28</td>\n",
       "      <td>19</td>\n",
       "      <td>26</td>\n",
       "      <td>53</td>\n",
       "      <td>43</td>\n",
       "      <td>19</td>\n",
       "      <td>66</td>\n",
       "      <td>93</td>\n",
       "      <td>39</td>\n",
       "      <td>84</td>\n",
       "      <td>85</td>\n",
       "      <td>46</td>\n",
       "      <td>40</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>v</th>\n",
       "      <td>50</td>\n",
       "      <td>41</td>\n",
       "      <td>71</td>\n",
       "      <td>16</td>\n",
       "      <td>60</td>\n",
       "      <td>33</td>\n",
       "      <td>17</td>\n",
       "      <td>33</td>\n",
       "      <td>69</td>\n",
       "      <td>44</td>\n",
       "      <td>22</td>\n",
       "      <td>78</td>\n",
       "      <td>124</td>\n",
       "      <td>52</td>\n",
       "      <td>102</td>\n",
       "      <td>85</td>\n",
       "      <td>54</td>\n",
       "      <td>45</td>\n",
       "      <td>15</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>y</th>\n",
       "      <td>32</td>\n",
       "      <td>18</td>\n",
       "      <td>42</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>33</td>\n",
       "      <td>31</td>\n",
       "      <td>11</td>\n",
       "      <td>41</td>\n",
       "      <td>65</td>\n",
       "      <td>27</td>\n",
       "      <td>56</td>\n",
       "      <td>44</td>\n",
       "      <td>31</td>\n",
       "      <td>20</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l</th>\n",
       "      <td>52</td>\n",
       "      <td>31</td>\n",
       "      <td>72</td>\n",
       "      <td>26</td>\n",
       "      <td>54</td>\n",
       "      <td>40</td>\n",
       "      <td>19</td>\n",
       "      <td>37</td>\n",
       "      <td>68</td>\n",
       "      <td>64</td>\n",
       "      <td>32</td>\n",
       "      <td>90</td>\n",
       "      <td>170</td>\n",
       "      <td>75</td>\n",
       "      <td>113</td>\n",
       "      <td>117</td>\n",
       "      <td>56</td>\n",
       "      <td>74</td>\n",
       "      <td>26</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i</th>\n",
       "      <td>45</td>\n",
       "      <td>26</td>\n",
       "      <td>52</td>\n",
       "      <td>11</td>\n",
       "      <td>35</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>35</td>\n",
       "      <td>43</td>\n",
       "      <td>21</td>\n",
       "      <td>49</td>\n",
       "      <td>114</td>\n",
       "      <td>41</td>\n",
       "      <td>73</td>\n",
       "      <td>89</td>\n",
       "      <td>51</td>\n",
       "      <td>38</td>\n",
       "      <td>6</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k</th>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>54</td>\n",
       "      <td>22</td>\n",
       "      <td>40</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>57</td>\n",
       "      <td>48</td>\n",
       "      <td>17</td>\n",
       "      <td>51</td>\n",
       "      <td>126</td>\n",
       "      <td>40</td>\n",
       "      <td>84</td>\n",
       "      <td>102</td>\n",
       "      <td>44</td>\n",
       "      <td>53</td>\n",
       "      <td>22</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>45</td>\n",
       "      <td>43</td>\n",
       "      <td>55</td>\n",
       "      <td>11</td>\n",
       "      <td>45</td>\n",
       "      <td>30</td>\n",
       "      <td>16</td>\n",
       "      <td>35</td>\n",
       "      <td>79</td>\n",
       "      <td>34</td>\n",
       "      <td>14</td>\n",
       "      <td>84</td>\n",
       "      <td>131</td>\n",
       "      <td>44</td>\n",
       "      <td>103</td>\n",
       "      <td>92</td>\n",
       "      <td>39</td>\n",
       "      <td>53</td>\n",
       "      <td>16</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>m</th>\n",
       "      <td>18</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>17</td>\n",
       "      <td>26</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "      <td>27</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h</th>\n",
       "      <td>18</td>\n",
       "      <td>12</td>\n",
       "      <td>24</td>\n",
       "      <td>9</td>\n",
       "      <td>19</td>\n",
       "      <td>15</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>39</td>\n",
       "      <td>22</td>\n",
       "      <td>30</td>\n",
       "      <td>39</td>\n",
       "      <td>18</td>\n",
       "      <td>16</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    g   d   f   w   r   q   c   n   a   e   p   t    v   y    l    i   k   s  \\\n",
       "g  67  51  69  25  42  37  23  24  79  58  30  77  156  59  114  114  64  63   \n",
       "d  43  41  62  18  43  24  14  24  62  39  22  56  139  47   84   88  38  42   \n",
       "f  23  21  43  14  31  15   9  19  33  30  15  37   65  33   56   53  33  28   \n",
       "w  13   8  18   4  11   4   6   3  11   9   3  14   21  13   22   13  12  11   \n",
       "r  30  28  31  17  45  22  10  17  40  56  16  64  106  47   91   64  35  35   \n",
       "q  28  16  25   9  32  12  12  20  36  27   8  51   81  33   60   56  23  30   \n",
       "c  14   5  12   4   4  12  11   7  15  11   4   7   41  10   16   15   8  11   \n",
       "n  38  20  35  18  42  30  14  22  45  28  11  56   74  35   65   78  40  24   \n",
       "a  63  29  62  20  44  29  17  37  95  58  22  89  178  59  131  133  48  54   \n",
       "e  53  30  59  19  45  36  19  32  63  43  21  64  147  42  104  101  49  46   \n",
       "p  40  15  43  13  36  23  14  19  49  34  19  45   81  38   65   50  37  40   \n",
       "t  47  28  44  13  41  28  19  26  53  43  19  66   93  39   84   85  46  40   \n",
       "v  50  41  71  16  60  33  17  33  69  44  22  78  124  52  102   85  54  45   \n",
       "y  32  18  42  12  21  20  17  16  33  31  11  41   65  27   56   44  31  20   \n",
       "l  52  31  72  26  54  40  19  37  68  64  32  90  170  75  113  117  56  74   \n",
       "i  45  26  52  11  35  23  13  18  35  43  21  49  114  41   73   89  51  38   \n",
       "k  43  43  54  22  40  23  20  19  57  48  17  51  126  40   84  102  44  53   \n",
       "s  45  43  55  11  45  30  16  35  79  34  14  84  131  44  103   92  39  53   \n",
       "m  18   5  11   5   7   3   6   5  12  10   7  17   26  16   21   27  15   9   \n",
       "h  18  12  24   9  19  15   4  11  15  17   5  20   39  22   30   39  18  16   \n",
       "\n",
       "    m   h  \n",
       "g  29  33  \n",
       "d  16  21  \n",
       "f  15   8  \n",
       "w   3   6  \n",
       "r   8  24  \n",
       "q   6  19  \n",
       "c   4   6  \n",
       "n  10  15  \n",
       "a  27  34  \n",
       "e  15  23  \n",
       "p  12  23  \n",
       "t  17  17  \n",
       "v  15  24  \n",
       "y   9  13  \n",
       "l  26  30  \n",
       "i   6  19  \n",
       "k  22  24  \n",
       "s  16  23  \n",
       "m   4   9  \n",
       "h   8   7  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj_aj_ajm_dict[\"b\"].iloc[:, :, 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
