{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Computational biology and bioinformatics - <span style=\"color:#1CA766\">INFO-F-439</span>\n",
    "# Assignment 1: <span style=\"color:#1CA766\">GOR III secondary structure prediction</span>\n",
    "\n",
    "> ## <span style=\"color:#2E66A7\"> Alberto Parravicini</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*****\n",
    "\n",
    "# <span style=\"color:#2E66A7\">Part 1:</span> Introduction\n",
    "\n",
    "Proteins are for the most part defined by the aminoacid sequences that compose them. However, due to the nature of the aminoacids, proteins have more complex structures that characterize them.\n",
    "\n",
    "Hydrogen bonds between the aminoacids in the sequence causes the proteins to assume three-dimensional structures of various kind. Being able to recognize these structures is useful to compare proteins, and to get a better insight on their properties.\n",
    "\n",
    "Classifying correctly which portion of a protein assumes a given secondary structure isn't an easy task, but it can be seen that there are links between specific patterns of aminoacids and given secondary structures.\n",
    "As an example, in $\\alpha$-helices (one type of secondary structure) every third aminoacid along the sequence will tend to be hydrophobic. Also, regions richer of *alanine (A)* and *leucine (L)* will tend to form helices. (see *Gollery, Martin. \"Bioinformatics: Sequence and Genome Analysis\", David W. Mount. Pag. 387-388*)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "### <span style=\"color:#1CA766\">GOR I & II</span>:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One idea that can be exploited to predict secondary structures from aminoacid sequeces, is that the secondary structure corresponding to a certain aminoacid is influenced not only by the aminoacid itself, but also by the aminoacids that are close to it.\n",
    "\n",
    "Given a dataset of proteins where for each aminoacid its secondary structure is known, it is possible to build a model that is able to predict secondary sequences, by using the idea above.\n",
    "\n",
    "This is the core idea behind the **GOR I** and **GOR II** algorithms.\n",
    "\n",
    "For a given position $j$, it will be predicted the secondary structure $S_j$ that has the highest value of *information*, as function of the aminoacid sequence.\n",
    "Assuming that our secondary structures are in the set $S = \\{\\alpha$-helices (**H**), $\\beta$-sheets (**B**), and coils (**C**)$\\}$, the one that will be chosen is $$S_j = \\arg\\max_{S}{I(\\Delta S_j;\\ a_1, \\ldots, a_n)}$$ where $a_1, \\ldots, a_n$ is the aminoacid sequence, and $S_j$ is picked among the secondary structures in S.\n",
    "\n",
    "The previous computation is approximated by **GOR** by considering a window of size $17$ around position $j$, which gives the formula $$S_j = \\arg\\max_{S}{I(\\Delta S_j;\\ a_1, \\ldots, a_n)} \\approx \\arg\\max_{S}{\\sum_{m = -8}^{m = 8}{I(\\Delta S_j;\\ a_{j+m})}}$$ \n",
    "\n",
    "In the previous formulation, $I(\\Delta S_j;\\ a_{j+m})$ is a measure of *self-information*, and is given by\n",
    "$$I(\\Delta S_j;\\ a_{j+m}) = log(\\frac{c_{S_j,\\ a_{j+m}}}{c_{\\neg S_j,\\ a_{j+m}}}) + log(\\frac{c_{\\neg S}}{c_{S}})$$\n",
    "\n",
    "where $c$ refers to the count of the occurrences (in a given dataset) of what is specified in the subscript; for instance, $c_{\\neg S}$ is the number of aminoacids with a secondary structure different from $S$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In Python, we can compute $c_{S_j,\\ a_{j+m}}$ and $I(\\Delta S_j;\\ a_{j+m})$ in the following way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def count_residues_gor_1(s_list, a_list, s_j, a_jm, lag):\n",
    "    \"\"\"\n",
    "    Count the number of times that structure \"s_j\" appears,\n",
    "    and residue \"a_jm\" is at distance \"lag\" from them.\n",
    "    :param s_list: a list of secondary structures; pandas Series.\n",
    "    :param a_list: a list of aminoacids; pandas Series.\n",
    "    :param s_j: a secondary structure name.\n",
    "    :param a_jm: an aminoacid name.\n",
    "    :param lag: distance from a_j at which a_jm is searched.\n",
    "    :return: int\n",
    "    \"\"\"\n",
    "    n = len(s_list)\n",
    "    count = 0\n",
    "    for i in range(max(0, -lag), min(n, n - lag)):\n",
    "        if s_list.iloc[i] == s_j and a_list.iloc[i + lag] == a_jm:\n",
    "            count += 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def info_value_gor_1(s_list, a_list, s_j, a_jm, lag, a_occ):\n",
    "    \"\"\"\n",
    "    Compute the information value associated to a secondary structure s_j, in position j,\n",
    "    and to an aminoacid a_jm, at distance m from s_j.\n",
    "    :param s_list: a list of secondary structures; pandas Series.\n",
    "    :param a_list: a list of aminoacids; pandas Series.\n",
    "    :param s_j: a secondary structure name.\n",
    "    :param a_jm: an aminoacid name.\n",
    "    :param lag: distance from a_j at which a_jm is searched.\n",
    "    :param a_occ: number of occurrencies of each aminoacid; pandas Series.\n",
    "    :return: double\n",
    "    \"\"\"\n",
    "    num_1 = count_residues_gor_1(s_list, a_list, s_j, a_jm, lag)\n",
    "    den_1 = a_occ[a_jm] - num_1\n",
    "    den_2 = s_occ[s_j]\n",
    "    num_2 = len(s_list) - den_2\n",
    "    return np.log(num_1 * num_2) - np.log(den_1 * den_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "### <span style=\"color:#1CA766\">GOR III</span>:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is possible to improve the accuracy of **GOR I & II** by considering not only the occurrencies of aminoacids close to the position where we want to predict the secondary structure, but also their occurrencies relatively to the aminoacid found at that given position.\n",
    "\n",
    "This means that the information value of a secondary structure $S_j$ will be approximated as \n",
    "\n",
    "$$S_j = \\arg\\max_{S}{I(\\Delta S_j;\\ a_1, \\ldots, a_n)} \\approx \\arg\\max_{S}{\\sum_{m = -8}^{m = 8}{I(\\Delta S_j;\\ a_{j+m}\\ |\\ a_j)}} = \\ldots$$ \n",
    "$$\\ldots = I(\\Delta S_j;\\ a_j) + \\sum_{m = -8,\\ m \\neq 0}^{m = 8}{I(\\Delta S_j;\\ a_{j+m}\\ |\\ a_j)}$$\n",
    "\n",
    "where it holds $$I(\\Delta S_j;\\ a_{j+m}\\ |\\ a_j) = log(\\frac{c_{S_j,\\ a_{j+m},\\ a_j}}{c_{\\neg S_j,\\ a_{j+m},\\ a_j}}) + log(\\frac{c_{\\neg S_j,\\ a_j}}{c_{S_j,\\ a_j}})$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In Python, we can compute $c_{S_j,\\ a_{j+m},\\ a_j}$ and $I(\\Delta S_j;\\ a_{j+m}\\ |\\ a_j)$ as:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def count_residues_gor_3(s_list, a_list, s_j, a_j, a_jm, lag):\n",
    "    \"\"\"\n",
    "    Count the number of times that structure \"s_j\" appears together with residue \"a_j\",\n",
    "    and residue \"a_jm\" is at distance \"lag\" from them.\n",
    "    :param s_list: a list of secondary structures; pandas Series.\n",
    "    :param a_list: a list of aminoacids; pandas Series.\n",
    "    :param s_j: a secondary structure name\n",
    "    :param a_j: an aminoacid name\n",
    "    :param a_jm: an aminoacid name\n",
    "    :param lag: distance from a_j at which a_jm is searched\n",
    "    :return: int\n",
    "    \"\"\"\n",
    "    n = len(s_list)\n",
    "    count = 0\n",
    "    for i in range(max(0, -lag), min(n, n - lag)):\n",
    "        if s_list.iat[i] == s_j and a_list.iat[i] == a_j and a_list.iat[i + lag] == a_jm:\n",
    "            count += 1\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def info_value_gor_3(s_list, a_list, s_j, a_j, a_jm, lag, a_occ):\n",
    "    \"\"\"\n",
    "    Compute the information value associated to an aminoacid a_j in position j,\n",
    "    with secondary structure s_j, in position j,\n",
    "    and to an aminoacid a_jm, at distance m from s_j.\n",
    "    :param s_list: a list of secondary structures; pandas Series.\n",
    "    :param a_list: a list of aminoacids; pandas Series.\n",
    "    :param s_j: a secondary structure name\n",
    "    :param a_j: an aminoacid name\n",
    "    :param a_jm: an aminoacid name\n",
    "    :param lag: distance from a_j at which a_jm is searched\n",
    "    :param a_occ: number of occurrencies of each aminoacid\n",
    "    :return: double\n",
    "    \"\"\"\n",
    "    # Num of times s_j appears with a_j, and a_jm is lag position distant.\n",
    "    num_1 = count_residues_gor_3(s_list, a_list, s_j, a_j, a_jm, lag)\n",
    "    # Num of times a_j appears with a secondary structure different from s_j,\n",
    "    # and a_jm is lag position distant.\n",
    "    den_1 = count_aminoacid_lag(a_list, a_j, a_jm, lag) - num_1\n",
    "    \n",
    "    # Number of times aminoacid a_j appears together with structure s_j.                        \n",
    "    den_2 = count_residues_gor_3(s_list, a_list, s_j, a_j, a_j, 0)                          \n",
    "    # Number of times aminoacid a_j appears with a different structure from s_j.\n",
    "    num_2 = a_occ[a_j] - den_2\n",
    "                 \n",
    "    info = np.log(num_1) - np.log(den_1) + np.log(num_2) - np.log(den_2)   \n",
    "    return info     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "*****\n",
    "\n",
    "# <span style=\"color:#2E66A7\">Part 2:</span> Data pre-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To compute the information values for each combination of aminoacids and secondary structures, it is necessary to have a large dataset of proteins with known secondary structures. We are provided with a set of 498 proteins, where the secondary structures have been determined by **DSSP** and **STRIDE**.\n",
    "\n",
    "Before being able to apply **GOR III**, it is required to preprocess the dataset, clean the inconsistencies that might be present, and store the aminoacid sequences in an appropriate way.\n",
    "\n",
    "It can be noted that there are a few rows with aminoacids that are not part of the usual 20. As the number of these rows is very limited compared to the total, we can just remove them entirely, and this won't affect the overall results of the algorithm."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#%% IMPORT DATA\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import timeit\n",
    "import pickle\n",
    "\n",
    "aminoacid_list = \"arndceqghilkmfpstwyv\"\n",
    "\n",
    "# Dictionary that maps aminoacid codes to single letters.\n",
    "aminoacid_codes = {\"ala\":  \"a\",\n",
    "                   \"arg\":  \"r\",\n",
    "                   \"asn\":  \"n\",\n",
    "                   \"asp\":  \"d\",\n",
    "                   \"cys\":  \"c\",\n",
    "                   \"gln\":  \"q\",\n",
    "                   \"glu\":  \"e\",\n",
    "                   \"gly\":  \"g\",\n",
    "                   \"his\":  \"h\",\n",
    "                   \"ile\":  \"i\",\n",
    "                   \"leu\":  \"l\",\n",
    "                   \"lys\":  \"k\",\n",
    "                   \"met\":  \"m\",\n",
    "                   \"phe\":  \"f\",\n",
    "                   \"pro\":  \"p\",\n",
    "                   \"ser\":  \"s\",\n",
    "                   \"thr\":  \"t\",\n",
    "                   \"trp\":  \"w\",\n",
    "                   \"tyr\":  \"y\",\n",
    "                   \"val\":  \"v\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def preprocess_input(file_name, aminoacid_codes):\n",
    "    \"\"\"\n",
    "    Load and preprocess a file containing protein sequences.\n",
    "    The output will be a DataFrame ready to be used by GOR.\n",
    "    :param file_name: string, name of the file to be opened.\n",
    "    :param aminoacid_codes: dictionary that maps aminoacid codes to single letters.\n",
    "    :return: DataFrame\n",
    "    \"\"\"\n",
    "    # Load the data\n",
    "    input_data = pd.read_csv(file_name,\n",
    "                         header=None, sep=\"\\t\",\n",
    "                         names=[\"PDB_code\", \"PDB_chain_code\", \"PDB_seq_code\", \"residue_name\", \"secondary_structure\"])\n",
    "\n",
    "    # CLEAN DATA - AMINOACIDS\n",
    "\n",
    "    # There are some values that are weird, like \"a\", \"b\", ...\n",
    "    # Some of those can be interpreted as correct aminoacids.\n",
    "\n",
    "    # Remove rows with \"X\", \"b\", \"UNK\"\n",
    "    # First, remove any leading spaces, then remove \"X\", \"b\", \"UNK\"\n",
    "    input_data[\"residue_name\"] = input_data[\"residue_name\"].str.strip()\n",
    "    input_data = input_data.drop(input_data[input_data[\"residue_name\"].isin([\"X\", \"b\", \"UNK\"])].index).reset_index(drop=True)\n",
    "\n",
    "\n",
    "    # Add a column for the aminoacids, while preserving the original one\n",
    "    input_data[\"a\"] = input_data[\"residue_name\"]\n",
    "    input_data[\"a\"] = input_data[\"a\"].str.lower()\n",
    "\n",
    "\n",
    "    # replace the codes with something shorter\n",
    "    input_data[\"a\"] = input_data[\"a\"].map(lambda x: aminoacid_codes[x] if x in aminoacid_codes else x)\n",
    "\n",
    "    # Create a new column\n",
    "    input_data[\"s\"] = input_data[\"secondary_structure\"]\n",
    "    input_data[\"s\"] = input_data[\"s\"].str.lower()\n",
    "    # Replace \"other\" with \"coil\"\n",
    "    input_data.loc[input_data[\"s\"] == \"other\", \"s\"] = \"coil\"\n",
    "    # Shorten values\n",
    "    input_data[\"s\"] = input_data[\"s\"].str[0]\n",
    "    \n",
    "    # Add a composite key to the dataset.\n",
    "    # Append the PDB chain code to the PDB code, to obtain unique protein identifier.\n",
    "    input_data[\"PDB_code_and_chain\"] = input_data.PDB_code + \"_\" + input_data.PDB_chain_code\n",
    "\n",
    "    return input_data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can load and work on the *STRIDE* dataset. The *DSSP* dataset is similar, and will be compared to *STRIDE* later in the report. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CYS', 'LYS', 'VAL', 'TYR', 'MET', 'PHE', 'ARG', 'SER', 'GLY', 'GLN', 'ASN', 'GLU', 'ALA', 'THR', 'PRO', 'HIS', 'ILE', 'ASP', 'LEU', 'TRP'}\n",
      "{'Other', 'Beta', 'Helix'}\n"
     ]
    }
   ],
   "source": [
    "    # Type of the data to read (\"stride\", \"dssp\")\n",
    "    data_type = \"stride\"\n",
    "    # File name\n",
    "    file_name = \"../data/\" + data_type + \"_info.txt\"\n",
    "    # Read the data\n",
    "    input_data = preprocess_input(file_name, aminoacid_codes)\n",
    "\n",
    "    # Look at the aminoacid values\n",
    "    aminoacids = set(input_data.residue_name)\n",
    "    print(aminoacids)\n",
    "    # Same stuff with the secondary structures\n",
    "    secondary_structures = set(input_data.secondary_structure)\n",
    "    print(secondary_structures)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PDB_code</th>\n",
       "      <th>PDB_chain_code</th>\n",
       "      <th>PDB_seq_code</th>\n",
       "      <th>residue_name</th>\n",
       "      <th>secondary_structure</th>\n",
       "      <th>a</th>\n",
       "      <th>s</th>\n",
       "      <th>PDB_code_and_chain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1w0n</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>ILE</td>\n",
       "      <td>Other</td>\n",
       "      <td>i</td>\n",
       "      <td>c</td>\n",
       "      <td>1w0n_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1w0n</td>\n",
       "      <td>A</td>\n",
       "      <td>13</td>\n",
       "      <td>THR</td>\n",
       "      <td>Beta</td>\n",
       "      <td>t</td>\n",
       "      <td>b</td>\n",
       "      <td>1w0n_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1w0n</td>\n",
       "      <td>A</td>\n",
       "      <td>14</td>\n",
       "      <td>LYS</td>\n",
       "      <td>Beta</td>\n",
       "      <td>k</td>\n",
       "      <td>b</td>\n",
       "      <td>1w0n_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1w0n</td>\n",
       "      <td>A</td>\n",
       "      <td>15</td>\n",
       "      <td>VAL</td>\n",
       "      <td>Beta</td>\n",
       "      <td>v</td>\n",
       "      <td>b</td>\n",
       "      <td>1w0n_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1w0n</td>\n",
       "      <td>A</td>\n",
       "      <td>16</td>\n",
       "      <td>GLU</td>\n",
       "      <td>Beta</td>\n",
       "      <td>e</td>\n",
       "      <td>b</td>\n",
       "      <td>1w0n_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PDB_code PDB_chain_code  PDB_seq_code residue_name secondary_structure  a  \\\n",
       "0     1w0n              A            12          ILE               Other  i   \n",
       "1     1w0n              A            13          THR                Beta  t   \n",
       "2     1w0n              A            14          LYS                Beta  k   \n",
       "3     1w0n              A            15          VAL                Beta  v   \n",
       "4     1w0n              A            16          GLU                Beta  e   \n",
       "\n",
       "   s PDB_code_and_chain  \n",
       "0  c             1w0n_A  \n",
       "1  b             1w0n_A  \n",
       "2  b             1w0n_A  \n",
       "3  b             1w0n_A  \n",
       "4  b             1w0n_A  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_data.iloc[0:5, ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**PDB_code** and **PDB_chain_code** work as identifier of the row. The column **a** and **s** are just shortened versions of the aminoacid codes and of the secondary structure codes.\n",
    "\n",
    "We can also compute some other values, that will be useful later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Number of residues\n",
    "n_res = len(input_data.index)\n",
    "\n",
    "# Compute the number of occurrencies of each aminoacid\n",
    "a_occ = input_data[\"a\"].value_counts()\n",
    "\n",
    "# Compute the number of occurrencies of each secondary structure\n",
    "s_occ = input_data[\"s\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "71073"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "l    6449\n",
       "a    5963\n",
       "g    5119\n",
       "v    4960\n",
       "e    4816\n",
       "d    4244\n",
       "k    4213\n",
       "s    4148\n",
       "i    4034\n",
       "t    3937\n",
       "r    3665\n",
       "p    3181\n",
       "n    3053\n",
       "f    2890\n",
       "q    2849\n",
       "y    2467\n",
       "h    1715\n",
       "m    1260\n",
       "c    1103\n",
       "w    1007\n",
       "Name: a, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a_occ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "c    29170\n",
       "h    25755\n",
       "b    16148\n",
       "Name: s, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_occ"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*****\n",
    "\n",
    "# <span style=\"color:#2E66A7\">Part 3:</span> GOR III"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, to compute the information values of **GOR III** it is required to add a few more functions.\n",
    "\n",
    "The first will compute a matrix that stores, for each aminoacid, the occurrences of each secondary structure.  \n",
    "The second will build a dictionary where each key corresponds to a secondary structure, and each value is a tensor (technically, a pandas.Panel) that stores the number of times where an aminoacid $a_j$ appears with secondary structure $S_j$,\n",
    "and another aminoacid $a_{j+m}$ appears at a distance $m$ from them, with $m$ in $[-8, 8]$.\n",
    "\n",
    "It is necessary to precompute these matrices as counting from scratch the occurrencies, for each prediction, could be very inefficient."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def build_sj_aj_matrix(s_list, a_list):\n",
    "    \"\"\"\n",
    "    Count the number of times that each aminoacid a_j has secondary structure s_j.\n",
    "    :param s_list: a list of secondary structures; pandas Series.\n",
    "    :param a_list: a list of aminoacids; pandas Series.\n",
    "    :return: pandas.DataFrame\n",
    "    \"\"\"\n",
    "    # Set of all secondary structures.\n",
    "    sec_structure_set = [\"h\", \"b\", \"c\"]\n",
    "    # Set of all aminoacids.\n",
    "    aminoacid_set = set(a_list)\n",
    "    # Occurrency matrix.\n",
    "    sj_aj_matrix = pd.DataFrame(0, index=sec_structure_set, columns=aminoacid_set, dtype=int)\n",
    "    # Zip together the lists of secondary structures and aminoacids, it makes counting easier.\n",
    "    s_a_list = [list(x) for x in zip(s_list,a_list)]\n",
    "    # Count the occurrencies.\n",
    "    for s in sec_structure_set:\n",
    "        for a in aminoacid_set:\n",
    "            sj_aj_matrix.at[s, a] = s_a_list.count([s, a])\n",
    "\n",
    "    return sj_aj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "! -> EXECUTION TIME OF build_sj_aj_matrix: 0.26465835296067575 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "start_time = timeit.default_timer()\n",
    "sj_aj_matrix = build_sj_aj_matrix(input_data.s, input_data.a)\n",
    "end_time = timeit.default_timer()\n",
    "print(\"! -> EXECUTION TIME OF build_sj_aj_matrix:\", (end_time - start_time), \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>i</th>\n",
       "      <th>v</th>\n",
       "      <th>n</th>\n",
       "      <th>l</th>\n",
       "      <th>f</th>\n",
       "      <th>c</th>\n",
       "      <th>t</th>\n",
       "      <th>p</th>\n",
       "      <th>a</th>\n",
       "      <th>s</th>\n",
       "      <th>g</th>\n",
       "      <th>e</th>\n",
       "      <th>k</th>\n",
       "      <th>h</th>\n",
       "      <th>q</th>\n",
       "      <th>y</th>\n",
       "      <th>r</th>\n",
       "      <th>d</th>\n",
       "      <th>w</th>\n",
       "      <th>m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>1572</td>\n",
       "      <td>2115</td>\n",
       "      <td>447</td>\n",
       "      <td>1573</td>\n",
       "      <td>938</td>\n",
       "      <td>301</td>\n",
       "      <td>1135</td>\n",
       "      <td>335</td>\n",
       "      <td>1007</td>\n",
       "      <td>771</td>\n",
       "      <td>804</td>\n",
       "      <td>773</td>\n",
       "      <td>809</td>\n",
       "      <td>400</td>\n",
       "      <td>493</td>\n",
       "      <td>827</td>\n",
       "      <td>742</td>\n",
       "      <td>530</td>\n",
       "      <td>294</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h</th>\n",
       "      <td>1468</td>\n",
       "      <td>1601</td>\n",
       "      <td>916</td>\n",
       "      <td>3071</td>\n",
       "      <td>1012</td>\n",
       "      <td>305</td>\n",
       "      <td>1089</td>\n",
       "      <td>467</td>\n",
       "      <td>2992</td>\n",
       "      <td>1204</td>\n",
       "      <td>850</td>\n",
       "      <td>2348</td>\n",
       "      <td>1795</td>\n",
       "      <td>570</td>\n",
       "      <td>1369</td>\n",
       "      <td>850</td>\n",
       "      <td>1586</td>\n",
       "      <td>1338</td>\n",
       "      <td>377</td>\n",
       "      <td>547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>994</td>\n",
       "      <td>1244</td>\n",
       "      <td>1690</td>\n",
       "      <td>1805</td>\n",
       "      <td>940</td>\n",
       "      <td>497</td>\n",
       "      <td>1713</td>\n",
       "      <td>2379</td>\n",
       "      <td>1964</td>\n",
       "      <td>2173</td>\n",
       "      <td>3465</td>\n",
       "      <td>1695</td>\n",
       "      <td>1609</td>\n",
       "      <td>745</td>\n",
       "      <td>987</td>\n",
       "      <td>790</td>\n",
       "      <td>1337</td>\n",
       "      <td>2376</td>\n",
       "      <td>336</td>\n",
       "      <td>431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      i     v     n     l     f    c     t     p     a     s     g     e  \\\n",
       "b  1572  2115   447  1573   938  301  1135   335  1007   771   804   773   \n",
       "h  1468  1601   916  3071  1012  305  1089   467  2992  1204   850  2348   \n",
       "c   994  1244  1690  1805   940  497  1713  2379  1964  2173  3465  1695   \n",
       "\n",
       "      k    h     q    y     r     d    w    m  \n",
       "b   809  400   493  827   742   530  294  282  \n",
       "h  1795  570  1369  850  1586  1338  377  547  \n",
       "c  1609  745   987  790  1337  2376  336  431  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj_aj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def build_sj_aj_ajm(input_data, print_details=True):\n",
    "    \"\"\"\n",
    "    Count the times where aminoacid a_j appears with secondary structure s_j,\n",
    "    and aminoacid a_jm appears at a distance m from them, with m in [-8, 8].\n",
    "    \"\"\"\n",
    "    # Set of all proteins.\n",
    "    protein_set = list(set(input_data.PDB_code_and_chain))\n",
    "    # Set of all secondary structures.\n",
    "    sec_structure_set =de [\"h\", \"b\", \"c\"]\n",
    "    # Set of all aminoacids.\n",
    "    aminoacid_set = list(set(input_data.a))\n",
    "    # Build a dictionary where the keys are the secondary structures, and the values are 3D tensors\n",
    "    # with index a_j, a_jm, m.\n",
    "    sj_aj_ajm_dict = {s: pd.Panel(data=0, items=aminoacid_set, major_axis=aminoacid_set, minor_axis=np.arange(-8, 9), dtype=int) for s in sec_structure_set}\n",
    "\n",
    "    # The counting must be done separately for each protein,\n",
    "    # otherwise the aminoacids at the end of one protein would be counted\n",
    "    # as part of the next one!\n",
    "    for i_p, p in enumerate(protein_set):\n",
    "        s_list = input_data.loc[input_data['PDB_code_and_chain'] == p].s\n",
    "        a_list = input_data.loc[input_data['PDB_code_and_chain'] == p].a\n",
    "        n = len(s_list)\n",
    "        for i_aj, a_j in enumerate(aminoacid_set):\n",
    "            for i_ajm, a_jm in enumerate(aminoacid_set):\n",
    "                for i_m, m in enumerate(np.arange(-8, 9)):\n",
    "                    if print_details:\n",
    "                        print(p, a_j, a_jm, m, \"-------\", i_p / len(protein_set), i_aj / 20, i_ajm / 20)\n",
    "\n",
    "                    for i in range(max(0, -m), min(n, n - m)):\n",
    "                        if a_list.iat[i] == a_j and a_list.iat[i + m] == a_jm:\n",
    "                            # Increment the count of the right dictionary.\n",
    "                            sj_aj_ajm_dict[s_list.iat[i]].iat[i_aj, i_ajm, i_m] += 1\n",
    "\n",
    "    return sj_aj_ajm_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "! -> EXECUTION TIME OF build_sj_aj_ajm: 5.784613011439177e-05 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Count the times where aminoacid a_j appears with secondary structure s_j,\n",
    "# and aminoacid a_jm appears at a distance m from them, with m in [-8, 8].\n",
    "#   WARNING: this can take a bit of time, usually about 20-30 minutes;\n",
    "#            do it at your own risk!\n",
    "start_time = timeit.default_timer()\n",
    "##########################\n",
    "# UNCOMMENT IF NEEDED ####\n",
    "##########################\n",
    "# sj_aj_ajm_dict = build_sj_aj_ajm(input_data)\n",
    "\n",
    "end_time = timeit.default_timer()\n",
    "print(\"! -> EXECUTION TIME OF build_sj_aj_ajm:\", (end_time - start_time), \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As the computation of the dictionary can take a fair bit of time, it is a good idea to store the result, and load it when needed.\n",
    "\n",
    "Note: comment/uncomment the following lines depending on your needs! The trained models should be already available, so there is no need to run the code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Save data\n",
    "dict_file_name = \"sj_aj_ajm_dict_\"+ data_type + \".p\"\n",
    "\n",
    "# # Open the file for writing.\n",
    "# file_object = open(dict_file_name.encode('utf-8').strip(), 'wb')\n",
    "\n",
    "# # Save data\n",
    "# pickle.dump(sj_aj_ajm_dict, file_object)\n",
    "#\n",
    "# file_object.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data can also be loaded with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Load data\n",
    "file_object = open(dict_file_name.encode('utf-8').strip(), 'rb')\n",
    "sj_aj_ajm_dict = pickle.load(file_object)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we can visualize a slice of **sj_aj_ajm_dict**: for instance, we can consider the occurrences of a $\\beta$-sheets with an aminoacid distance of $-1$.  \n",
    "Among other things, we can notice how **v** is often followed by another **v**, while **p** is followed by another **p** just one time!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g</th>\n",
       "      <th>d</th>\n",
       "      <th>f</th>\n",
       "      <th>w</th>\n",
       "      <th>r</th>\n",
       "      <th>q</th>\n",
       "      <th>c</th>\n",
       "      <th>n</th>\n",
       "      <th>a</th>\n",
       "      <th>e</th>\n",
       "      <th>p</th>\n",
       "      <th>t</th>\n",
       "      <th>v</th>\n",
       "      <th>y</th>\n",
       "      <th>l</th>\n",
       "      <th>i</th>\n",
       "      <th>k</th>\n",
       "      <th>s</th>\n",
       "      <th>m</th>\n",
       "      <th>h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>g</th>\n",
       "      <td>50</td>\n",
       "      <td>53</td>\n",
       "      <td>67</td>\n",
       "      <td>28</td>\n",
       "      <td>60</td>\n",
       "      <td>47</td>\n",
       "      <td>17</td>\n",
       "      <td>25</td>\n",
       "      <td>48</td>\n",
       "      <td>67</td>\n",
       "      <td>13</td>\n",
       "      <td>87</td>\n",
       "      <td>139</td>\n",
       "      <td>70</td>\n",
       "      <td>96</td>\n",
       "      <td>90</td>\n",
       "      <td>70</td>\n",
       "      <td>63</td>\n",
       "      <td>17</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>34</td>\n",
       "      <td>10</td>\n",
       "      <td>32</td>\n",
       "      <td>10</td>\n",
       "      <td>25</td>\n",
       "      <td>19</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>34</td>\n",
       "      <td>29</td>\n",
       "      <td>8</td>\n",
       "      <td>35</td>\n",
       "      <td>70</td>\n",
       "      <td>43</td>\n",
       "      <td>58</td>\n",
       "      <td>61</td>\n",
       "      <td>37</td>\n",
       "      <td>26</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f</th>\n",
       "      <td>49</td>\n",
       "      <td>32</td>\n",
       "      <td>34</td>\n",
       "      <td>8</td>\n",
       "      <td>45</td>\n",
       "      <td>33</td>\n",
       "      <td>12</td>\n",
       "      <td>23</td>\n",
       "      <td>56</td>\n",
       "      <td>57</td>\n",
       "      <td>20</td>\n",
       "      <td>61</td>\n",
       "      <td>99</td>\n",
       "      <td>50</td>\n",
       "      <td>55</td>\n",
       "      <td>76</td>\n",
       "      <td>56</td>\n",
       "      <td>68</td>\n",
       "      <td>12</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>w</th>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>24</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "      <td>29</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r</th>\n",
       "      <td>43</td>\n",
       "      <td>24</td>\n",
       "      <td>71</td>\n",
       "      <td>15</td>\n",
       "      <td>34</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>40</td>\n",
       "      <td>26</td>\n",
       "      <td>21</td>\n",
       "      <td>43</td>\n",
       "      <td>114</td>\n",
       "      <td>39</td>\n",
       "      <td>92</td>\n",
       "      <td>80</td>\n",
       "      <td>28</td>\n",
       "      <td>29</td>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>q</th>\n",
       "      <td>33</td>\n",
       "      <td>7</td>\n",
       "      <td>33</td>\n",
       "      <td>9</td>\n",
       "      <td>17</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>37</td>\n",
       "      <td>17</td>\n",
       "      <td>7</td>\n",
       "      <td>30</td>\n",
       "      <td>82</td>\n",
       "      <td>24</td>\n",
       "      <td>70</td>\n",
       "      <td>53</td>\n",
       "      <td>24</td>\n",
       "      <td>23</td>\n",
       "      <td>8</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "      <td>24</td>\n",
       "      <td>19</td>\n",
       "      <td>23</td>\n",
       "      <td>26</td>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n</th>\n",
       "      <td>25</td>\n",
       "      <td>10</td>\n",
       "      <td>31</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>25</td>\n",
       "      <td>22</td>\n",
       "      <td>9</td>\n",
       "      <td>41</td>\n",
       "      <td>83</td>\n",
       "      <td>21</td>\n",
       "      <td>46</td>\n",
       "      <td>57</td>\n",
       "      <td>28</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>42</td>\n",
       "      <td>25</td>\n",
       "      <td>64</td>\n",
       "      <td>13</td>\n",
       "      <td>50</td>\n",
       "      <td>26</td>\n",
       "      <td>25</td>\n",
       "      <td>32</td>\n",
       "      <td>91</td>\n",
       "      <td>61</td>\n",
       "      <td>21</td>\n",
       "      <td>75</td>\n",
       "      <td>136</td>\n",
       "      <td>57</td>\n",
       "      <td>99</td>\n",
       "      <td>120</td>\n",
       "      <td>38</td>\n",
       "      <td>36</td>\n",
       "      <td>16</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>e</th>\n",
       "      <td>32</td>\n",
       "      <td>23</td>\n",
       "      <td>38</td>\n",
       "      <td>17</td>\n",
       "      <td>39</td>\n",
       "      <td>16</td>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "      <td>51</td>\n",
       "      <td>38</td>\n",
       "      <td>15</td>\n",
       "      <td>43</td>\n",
       "      <td>117</td>\n",
       "      <td>46</td>\n",
       "      <td>88</td>\n",
       "      <td>86</td>\n",
       "      <td>37</td>\n",
       "      <td>33</td>\n",
       "      <td>12</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>29</td>\n",
       "      <td>10</td>\n",
       "      <td>19</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>25</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>86</td>\n",
       "      <td>12</td>\n",
       "      <td>55</td>\n",
       "      <td>58</td>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>t</th>\n",
       "      <td>55</td>\n",
       "      <td>27</td>\n",
       "      <td>73</td>\n",
       "      <td>25</td>\n",
       "      <td>31</td>\n",
       "      <td>32</td>\n",
       "      <td>26</td>\n",
       "      <td>24</td>\n",
       "      <td>73</td>\n",
       "      <td>23</td>\n",
       "      <td>27</td>\n",
       "      <td>69</td>\n",
       "      <td>153</td>\n",
       "      <td>63</td>\n",
       "      <td>130</td>\n",
       "      <td>120</td>\n",
       "      <td>36</td>\n",
       "      <td>45</td>\n",
       "      <td>29</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>v</th>\n",
       "      <td>108</td>\n",
       "      <td>81</td>\n",
       "      <td>83</td>\n",
       "      <td>34</td>\n",
       "      <td>93</td>\n",
       "      <td>44</td>\n",
       "      <td>33</td>\n",
       "      <td>56</td>\n",
       "      <td>133</td>\n",
       "      <td>103</td>\n",
       "      <td>53</td>\n",
       "      <td>158</td>\n",
       "      <td>233</td>\n",
       "      <td>103</td>\n",
       "      <td>185</td>\n",
       "      <td>162</td>\n",
       "      <td>104</td>\n",
       "      <td>79</td>\n",
       "      <td>31</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>y</th>\n",
       "      <td>33</td>\n",
       "      <td>28</td>\n",
       "      <td>59</td>\n",
       "      <td>13</td>\n",
       "      <td>43</td>\n",
       "      <td>37</td>\n",
       "      <td>13</td>\n",
       "      <td>26</td>\n",
       "      <td>52</td>\n",
       "      <td>33</td>\n",
       "      <td>11</td>\n",
       "      <td>49</td>\n",
       "      <td>64</td>\n",
       "      <td>33</td>\n",
       "      <td>70</td>\n",
       "      <td>73</td>\n",
       "      <td>39</td>\n",
       "      <td>43</td>\n",
       "      <td>9</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l</th>\n",
       "      <td>63</td>\n",
       "      <td>53</td>\n",
       "      <td>76</td>\n",
       "      <td>22</td>\n",
       "      <td>72</td>\n",
       "      <td>53</td>\n",
       "      <td>19</td>\n",
       "      <td>44</td>\n",
       "      <td>73</td>\n",
       "      <td>74</td>\n",
       "      <td>24</td>\n",
       "      <td>136</td>\n",
       "      <td>210</td>\n",
       "      <td>67</td>\n",
       "      <td>135</td>\n",
       "      <td>118</td>\n",
       "      <td>74</td>\n",
       "      <td>84</td>\n",
       "      <td>28</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>i</th>\n",
       "      <td>79</td>\n",
       "      <td>76</td>\n",
       "      <td>70</td>\n",
       "      <td>20</td>\n",
       "      <td>69</td>\n",
       "      <td>43</td>\n",
       "      <td>24</td>\n",
       "      <td>61</td>\n",
       "      <td>102</td>\n",
       "      <td>80</td>\n",
       "      <td>41</td>\n",
       "      <td>109</td>\n",
       "      <td>160</td>\n",
       "      <td>60</td>\n",
       "      <td>121</td>\n",
       "      <td>131</td>\n",
       "      <td>90</td>\n",
       "      <td>80</td>\n",
       "      <td>21</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k</th>\n",
       "      <td>38</td>\n",
       "      <td>15</td>\n",
       "      <td>42</td>\n",
       "      <td>11</td>\n",
       "      <td>28</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>22</td>\n",
       "      <td>47</td>\n",
       "      <td>44</td>\n",
       "      <td>28</td>\n",
       "      <td>47</td>\n",
       "      <td>139</td>\n",
       "      <td>32</td>\n",
       "      <td>90</td>\n",
       "      <td>94</td>\n",
       "      <td>38</td>\n",
       "      <td>37</td>\n",
       "      <td>25</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s</th>\n",
       "      <td>44</td>\n",
       "      <td>21</td>\n",
       "      <td>73</td>\n",
       "      <td>18</td>\n",
       "      <td>25</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "      <td>22</td>\n",
       "      <td>47</td>\n",
       "      <td>25</td>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>96</td>\n",
       "      <td>41</td>\n",
       "      <td>74</td>\n",
       "      <td>90</td>\n",
       "      <td>23</td>\n",
       "      <td>31</td>\n",
       "      <td>14</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>m</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "      <td>23</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "      <td>34</td>\n",
       "      <td>9</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>25</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h</th>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>26</td>\n",
       "      <td>11</td>\n",
       "      <td>22</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>25</td>\n",
       "      <td>16</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>48</td>\n",
       "      <td>23</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>13</td>\n",
       "      <td>25</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     g   d   f   w   r   q   c   n    a    e   p    t    v    y    l    i  \\\n",
       "g   50  53  67  28  60  47  17  25   48   67  13   87  139   70   96   90   \n",
       "d   34  10  32  10  25  19  13   9   34   29   8   35   70   43   58   61   \n",
       "f   49  32  34   8  45  33  12  23   56   57  20   61   99   50   55   76   \n",
       "w   13   5   7   4  19  11   6  12   13   24   2   18   28   15   29   20   \n",
       "r   43  24  71  15  34  13  12  21   40   26  21   43  114   39   92   80   \n",
       "q   33   7  33   9  17  12   8  13   37   17   7   30   82   24   70   53   \n",
       "c   19   9  12   5  12  16  14   5   19   12   7   18   24   19   23   26   \n",
       "n   25  10  31  15  16   8  12   6   25   22   9   41   83   21   46   57   \n",
       "a   42  25  64  13  50  26  25  32   91   61  21   75  136   57   99  120   \n",
       "e   32  23  38  17  39  16  13  22   51   38  15   43  117   46   88   86   \n",
       "p   19  10  29  10  19  17   9   7   25   10   1   22   86   12   55   58   \n",
       "t   55  27  73  25  31  32  26  24   73   23  27   69  153   63  130  120   \n",
       "v  108  81  83  34  93  44  33  56  133  103  53  158  233  103  185  162   \n",
       "y   33  28  59  13  43  37  13  26   52   33  11   49   64   33   70   73   \n",
       "l   63  53  76  22  72  53  19  44   73   74  24  136  210   67  135  118   \n",
       "i   79  76  70  20  69  43  24  61  102   80  41  109  160   60  121  131   \n",
       "k   38  15  42  11  28  20  17  22   47   44  28   47  139   32   90   94   \n",
       "s   44  21  73  18  25  18  15  22   47   25   7   57   96   41   74   90   \n",
       "m    9   9  18   6  23  15   3   9   16   12   7   18   34    9   21   21   \n",
       "h   16  12  26  11  22  13  10   8   25   16  13   19   48   23   36   36   \n",
       "\n",
       "     k   s   m   h  \n",
       "g   70  63  17  22  \n",
       "d   37  26   7  13  \n",
       "f   56  68  12  27  \n",
       "w   21  13   7  11  \n",
       "r   28  29   8  23  \n",
       "q   24  23   8  17  \n",
       "c   13  14   3   9  \n",
       "n   28  18   8   9  \n",
       "a   38  36  16  24  \n",
       "e   37  33  12  17  \n",
       "p   15  12  11   9  \n",
       "t   36  45  29  18  \n",
       "v  104  79  31  46  \n",
       "y   39  43   9  24  \n",
       "l   74  84  28  45  \n",
       "i   90  80  21  36  \n",
       "k   38  37  25  14  \n",
       "s   23  31  14  19  \n",
       "m   25  12   7   7  \n",
       "h   13  25   9  10  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sj_aj_ajm_dict[\"b\"].iloc[:, :, 7]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "*****\n",
    "\n",
    "# <span style=\"color:#2E66A7\">Part 4:</span> Secondary structure prediction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the previous section, the algorithm was trained by considering all the data at our disposal. However, it is usually a good idea to employ a training set that contains different data from the one for which a prediction will be made.  \n",
    "If not, our model will perform really well on the data at our disposal, but will fail to generalize to new, unseen data.  \n",
    "\n",
    "To be able to predict the secondary structure for all our aminoacids, but at the same time use as many data as possible in training, we can employ the so-called **leave-one-out** strategy: we chose the aminoacid for which we want to make the prediction, and take it out of the dataset; then, we train the algorithm on all the remaining data, and use that model to predict the secondary structure of the aminoacid we removed.  \n",
    "\n",
    "This technique allows to have a good estimate of what will be the performances of our model when it will be employed against new data, and is commonly used when few data are available, and training the model can be done quickly enough (as we have to build a new model for each of our elements!)\n",
    "\n",
    "To implement the **leave-one-out** strategy, we have to decide what should be removed from the training set: one obvious choice is to just take out from the dataset the aminoacid for which we make the prediction. However, we have more than $70000$ aminoacids, and taking one out won't affect the model in any significant way.  \n",
    "On the other hand, we can remove the entire protein to which that aminoacid belongs. This approach works better, as we take out from the model all the information that are somehow connected to our aminoacid.\n",
    "\n",
    "Now, it should be noted that rebuilding the model from scratch for each aminoacid will take too much time. Instead, we can leverage the model that we built on the entire dataset, and subtract to it the values that are related to the aminoacid that we want to take out."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "First, it is a good idea to build an auxiliary data-structure that will hold the results of the predictions, so it will be easy to perform further analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class SecStructPrediction:\n",
    "    \"\"\"\n",
    "    Class used to store the secondary structure predictions of proteins.\n",
    "    It can store the PDB code of a protein, its aminoacid sequence, the real secondary structure and the predicted one,\n",
    "    and more. Fields can be left empty if the information are not available.\n",
    "    \"\"\"\n",
    "    def __init__(self, pdb_code=\"\", protein=\"\", prediction=\"\", real_sec_structure=\"\", q3=0, mcc_h=0, mcc_b=0, mcc_c=0, mcc=0, overall_prediction=\"\"):\n",
    "        self.pdb_code = pdb_code\n",
    "        self.protein = protein\n",
    "        self.prediction = prediction\n",
    "        self.real_sec_structure = real_sec_structure\n",
    "        self.q3 = q3\n",
    "        self.mcc_h = mcc_h\n",
    "        self.mcc_b = mcc_b\n",
    "        self.mcc_c = mcc_c\n",
    "        self.mcc = mcc\n",
    "        self.overall_prediction = overall_prediction\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.pdb_code + \":\\n\" +\\\n",
    "               self.protein + \"\\n\" +\\\n",
    "               self.real_sec_structure + \"\\n\" +\\\n",
    "               self.prediction +\\\n",
    "               \"\\nMETRICS -- Q3:\" + \"{:.3f}\".format(self.q3) +\\\n",
    "               \" -- MCC_h:\" + \"{:.3f}\".format(self.mcc_h) +\\\n",
    "               \" -- MCC_b:\" + \"{:.3f}\".format(self.mcc_b) +\\\n",
    "               \" -- MCC_c:\" + \"{:.3f}\".format(self.mcc_c) +\\\n",
    "               \" -- MCC:\" + \"{:.3f}\".format(self.mcc) +\\\n",
    "               \" -- TOT. PRED.:\" + self.overall_prediction + \"\\n\"\n",
    "\n",
    "    def __repr__(self):\n",
    "        return str(self)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "Then, compute the **GOR III** information value given some aminoacid for which we want to make the prediction, and the model we trained earlier.  \n",
    "Note that it is possible to do **leave-one-out** by setting the appropriate parameter.\n",
    "\n",
    "This function will return the prediction of the most suitable secondary structure for the given aminoacid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_gor_3(a_list, i, sj_aj_ajm_dict, sj_aj_matrix, a_occ, l_1_out=False):\n",
    "    \"\"\"\n",
    "    Compute the GOR III information value for the given aminoacid, and return the best secondary structure prediction\n",
    "    :param a_list: pandas.Series; input aminoacid sequence.\n",
    "    :param i: position in the sequence for which the prediction is done.\n",
    "    :param sj_aj_ajm_dict: dictionary used by the GOR III model.\n",
    "    :param sj_aj_matrix: occurrency matrix used by the GOR III model.\n",
    "    :param a_occ: occurrencies of each aminoacid, used by the GOR III model.\n",
    "    :param l_1_out: boolean; do leave one out on the given GOR III model.\n",
    "    :return: a tuple containing (structure_prediction, prediction_score)\n",
    "    \"\"\"\n",
    "    if i < 0 or i >= len(a_list):\n",
    "        print(\"Index out of range:\", i, \"-- List length:\", len(a_list))\n",
    "        raise IndexError\n",
    "    \n",
    "    # The range in GOR is [-7, 7], but it might be smaller\n",
    "    # for the aminoacid at the start or end of the protein.\n",
    "    min_range = max(-i, -8)\n",
    "    max_range = min(9, len(a_list) - i)\n",
    "    \n",
    "    a_j = a_list.iat[i]\n",
    "\n",
    "\n",
    "    info_tot = {\"h\": 0, \"b\": 0, \"c\": 0}\n",
    "    for s in info_tot.keys():\n",
    "        for m in range(min_range, max_range):\n",
    "            a_jm = a_list.iat[i + m]\n",
    "\n",
    "            # Values to subtract in \"leave-one-out\"\n",
    "            sub_1 = (1 if l_1_out else 0)\n",
    "            sub_2 = (1 if l_1_out else 0)\n",
    "\n",
    "            # Num of times s_j appears with a_j, and a_jm is lag position distant.\n",
    "            num_1 = sj_aj_ajm_dict[s].at[a_j, a_jm, m] - sub_1\n",
    "            # Num of times a_j appears with a secondary structure different from s_j,\n",
    "            # and a_jm is lag position distant.\n",
    "            den_1 = sum([sj_aj_ajm_dict[s_i].at[a_j, a_jm, m] for s_i in info_tot.keys()]) - num_1 - sub_1\n",
    "\n",
    "            # Number of times aminoacid a_j appears together with structure s_j.\n",
    "            den_2 = sj_aj_matrix.at[s, a_j] - sub_2\n",
    "            # Number of times aminoacid a_j appears with a different structure from s_j.\n",
    "            num_2 = a_occ.at[a_j] - den_2 - sub_2\n",
    "\n",
    "            # Add up the information value.\n",
    "            info_tot[s] += np.log(num_1) - np.log(den_1) + np.log(num_2) - np.log(den_2)\n",
    "\n",
    "    # Predict the secondary structure with highest value.\n",
    "    return [max(info_tot, key=info_tot.get), max(info_tot.values())]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "To evaluate the qualities of the predictions, we can use different metrics: the simplest one is the **Q3** score, which measures the accuracy of our predictions, defined as the percentage of secondary structures that were correctly predicted.\n",
    "\n",
    "$$Q3 = \\frac{N_{correct}}{N_{total}}$$\n",
    "\n",
    "Another, more robust, metric, is the **MCC**, which is computed for each of the secondary structures. After considering a target secondary structure (e.g. $\\beta$-sheet), this metric takes into account not only how many structures we have predicted correctly, but also penalizes *false positives/negatives* guesses.  \n",
    "\n",
    "\n",
    "$$MCC = \\frac{TP\\cdot TN - FP\\cdot FN}{\\sqrt{(TP+FP)(TP + FN)(TN+FP)(TN+FN)}}$$\n",
    "\n",
    "where $TP, TN, FP, FN$ are the number of *true positives, true negatives, false positives, false negatives*."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "**MCC** can be implemented as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compute_mcc(prediction_seq, real_seq, sec_structure):\n",
    "    \"\"\"\n",
    "    Compute the MCC for a given secondary sequence prediction.\n",
    "    :param prediction_seq: string, the sequence of predicted secondary structures.\n",
    "    :param real_seq: string, the sequence of real secondary structures.\n",
    "    :param sec_structure: string, the secondary structure to use as target (usually one of \"h\", \"b\", \"c\")\n",
    "    :return: double\n",
    "    \"\"\"\n",
    "    tp = sum([prediction_seq[i] == sec_structure and real_seq[i] == sec_structure for i in range(len(prediction_seq))])\n",
    "    fn = sum([prediction_seq[i] != sec_structure and real_seq[i] == sec_structure for i in range(len(prediction_seq))])\n",
    "    fp = sum([prediction_seq[i] == sec_structure and real_seq[i] != sec_structure for i in range(len(prediction_seq))])\n",
    "    tn = sum([prediction_seq[i] != sec_structure and real_seq[i] != sec_structure for i in range(len(prediction_seq))])\n",
    "    # If all the values at denominator are different from 0, use the standard formula;\n",
    "    if (tp + fp)*(tp + fn)*(tn + fp)*(tn + fn) != 0:\n",
    "        mcc = (tp*tn - fp*fn) / np.sqrt((tp + fp)*(tp + fn)*(tn + fp)*(tn + fn))\n",
    "    else:\n",
    "        # If one value is equal to 0, set the denominator to 1.\n",
    "        # See https://en.wikipedia.org/wiki/Matthews_correlation_coefficient\n",
    "        # Doing so will give an mcc af 0, i.e. a null correlation.\n",
    "        mcc = tp*tn - fp*fn\n",
    "    return mcc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "Now, we can apply the prediction function to all the aminoacid in the dataset, and obtain their predicitions.  \n",
    "Moreover, we compute the **Q3** (accuracy) and **MCC** scores, and save those values too.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def predict_sec_structures(input_data, data_type=\"dssp\", print_details=False):\n",
    "    \"\"\"\n",
    "    Predict the secondary structures of the input data set.\n",
    "    The input data set is divided in proteins, and for each protein are given\n",
    "    the secondary structure prediction, and accuracy metrics.\n",
    "    The results are stored in a file.\n",
    "    :param input_data: pandas.DataFrame, the input sequence of proteins.\n",
    "    :param data_type: string, the source of the input data, \"dssp\" or \"stride\"\n",
    "    :param print_details: boolean, print the details of the prediction.\n",
    "    :return: list of SecStructPrediction.\n",
    "    \"\"\"\n",
    "\n",
    "    # Load data\n",
    "    dict_file_name = \"sj_aj_ajm_dict_\"+ data_type + \".p\"\n",
    "    # Load the model\n",
    "    file_object = open(dict_file_name.encode('utf-8').strip(), 'rb')\n",
    "    sj_aj_ajm_dict = pickle.load(file_object)\n",
    "\n",
    "    # Compute sj_aj_matrix abd a_occ;\n",
    "    # it's done again as it's fast, no need to store the precomputed matrices.\n",
    "    a_occ = input_data[\"a\"].value_counts()\n",
    "    sj_aj_matrix = build_sj_aj_matrix(input_data.s, input_data.a)\n",
    "\n",
    "    # List that contains all the predictions, divided by protein.\n",
    "    prediction_list = []\n",
    "    # Measure the overall accuracy.\n",
    "    tot_acc = 0\n",
    "    tot_count = 0\n",
    "    for p_i, p in enumerate(list(set(input_data.PDB_code ))):\n",
    "        # Measure the accuracy over a single protein.\n",
    "        acc = 0\n",
    "        pred = \"\"\n",
    "        # Extract aminoacid and secondary structures for a given protein.\n",
    "        a_list = input_data.loc[input_data['PDB_code'] == p].a\n",
    "        s_list = input_data.loc[input_data['PDB_code'] == p].s\n",
    "\n",
    "        for i in range(0, len(a_list)):\n",
    "            # Predict the secondary structure for a given aminoacid of the protein.\n",
    "            res = compute_gor_3(a_list, i, sj_aj_ajm_dict, sj_aj_matrix, a_occ, l_1_out=True)\n",
    "            # Add the new secondary structure that is predicted\n",
    "            pred += res[0]\n",
    "            if res[0] == s_list.iat[i]:\n",
    "                acc += 1\n",
    "\n",
    "        mcc_h=compute_mcc(pred, s_list.to_string(header=False, index=False).replace(\"\\n\", \"\"), \"h\")\n",
    "        mcc_b=compute_mcc(pred, s_list.to_string(header=False, index=False).replace(\"\\n\", \"\"), \"b\")\n",
    "        mcc_c=compute_mcc(pred, s_list.to_string(header=False, index=False).replace(\"\\n\", \"\"), \"c\")\n",
    "\n",
    "        prediction = SecStructPrediction(pdb_code=p,\n",
    "                                         protein=a_list.to_string(header=False, index=False).replace(\"\\n\", \"\"),\n",
    "                                         prediction=pred,\n",
    "                                         real_sec_structure=s_list.to_string(header=False, index=False).replace(\"\\n\", \"\"),\n",
    "                                         q3=acc / len(a_list),\n",
    "                                         mcc_h=mcc_h,\n",
    "                                         mcc_b=mcc_b,\n",
    "                                         mcc_c=mcc_c,\n",
    "                                         mcc=(mcc_h + mcc_b + mcc_c)/3\n",
    "                                         )\n",
    "\n",
    "        prediction_list.append(prediction)\n",
    "        tot_count += len(a_list)\n",
    "        tot_acc += acc\n",
    "        # Print the current prediction and the cumulated metrics.\n",
    "        if print_details:\n",
    "            print(prediction)\n",
    "            print(\"(\", p_i + 1,\"/\", len(list(set(input_data.PDB_code))), \") -- TOTAL ACCURACY:\", \"{:.3f}\".format(tot_acc / tot_count))\n",
    "\n",
    "    return prediction_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "Finally, we can compute the predictions for all our aminoacids, and store the results in a *csv* file that will be analyzed later.\n",
    "Alongside the predictions and the scores, we save additional information, such as the number of predicted secondary structures for each protein, and the number of each aminoacids in the protein. These information will be useful when predicting the overall secondary structure of the protein.\n",
    "\n",
    "Note that computing all the predictions can also take a bit of time!\n",
    "As an example, here the prediction is done just for the **STRIDE** dataset, but the results for both datasets have been already precomputed. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2x1f_A:\n",
      "psrvvylgsipydqteeqildlcsnvgpvinlkmmfdpqtgrskgyafiefrdlessasavrnlngyqlgsrflkcgyssnsdisgvslehhhh\n",
      "cccbbbbbccccccchhhhhhhhhhhccbbbbbbbbccccccbbbbbbbbbcchhhhhhhhhhhccbbbccbbbbbbbccccccchhhhhhhcc\n",
      "ccbbbbbbbcccbcchhhhhhchhcchbbbbbbbbbcccccchcbbbbbbhbbhhhchhhhhhhchhbbhhbbbbbbccccbcccbcbbbhchc\n",
      "METRICS -- Q3:0.713 -- MCC_h:0.543 -- MCC_b:0.643 -- MCC_c:0.522 -- MCC:0.570 -- TOT. PRED.:\n",
      "\n",
      "( 1 / 498 ) -- TOTAL ACCURACY: 0.713\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\albyr\\Anaconda3\\lib\\site-packages\\ipykernel\\__main__.py:45: RuntimeWarning: divide by zero encountered in log\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3mxz_A:\n",
      "hmatirnlkiktstckrivkelhsyekevereaaktadmkdkgadpydlkqqenvlgesrmmipdchkrlesaladlkstlaeleetekegpeiedakktvadvekq\n",
      "chhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhccchhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhcccchhhhhhhhhhhhhhhc\n",
      "hhhhhbhhccchhhhhhhhhbhchhchhhhhhhhhhhhhhhhccchchhhhhhhhhhchhbbhchhhhhhhhhhhhhhhhhhhhhhhchcchhhhhhhhhhhhhhhc\n",
      "METRICS -- Q3:0.869 -- MCC_h:0.476 -- MCC_b:0.000 -- MCC_c:0.557 -- MCC:0.344 -- TOT. PRED.:\n",
      "\n",
      "( 2 / 498 ) -- TOTAL ACCURACY: 0.796\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\albyr\\Anaconda3\\lib\\site-packages\\ipykernel\\__main__.py:45: RuntimeWarning: invalid value encountered in log\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1od6_A:\n",
      "mhvvypgsfdpltnghldviqrasrlfekvtvavlenqylfsaeerlaiireatahlanveaatfsgllvdfvrrvgaqaivkglravsdyeyelqmahlnrqlypgletlfilaatrysfvsstmvkeiaryggdvsklvppatlralkaklgq\n",
      "cbbbbbbcccccchhhhhhhhhhhhhccbbbbbbbccccccchhhhhhhhhhhhcccccbbbbbbccchhhhhhhhcccbbbbbbccccchhhhhhhhhhhhhhccccbbbbbbccccccccchhhhhhhhhhccccccccchhhhhhhhhhhcc\n",
      "bbbbcbccccbccchbhhhhhhhhhhhhhhhbbhhhhhhhhhhhhhhhhhhhhhhhhhhhbhhhhbcchbhhhhhhhhhhhhbhcbhhchhhhhhhhhhhhhhhcchhchbhhhhhhbbhbhhhhhhhbhbhhccccccbcbhchhhhhhhhhhh\n",
      "METRICS -- Q3:0.568 -- MCC_h:0.394 -- MCC_b:0.213 -- MCC_c:0.328 -- MCC:0.312 -- TOT. PRED.:\n",
      "\n",
      "( 3 / 498 ) -- TOTAL ACCURACY: 0.697\n",
      "1usm_A:\n",
      "mdweerkrlvktfafpnfrealdfanrvgalaerenhhprltvewgrvtvewwthsaggvtekdremarltdallqr\n",
      "cccbbcccbbbbbbccchhhhhhhhhhhhhhhhhhccccbbbbbccbbbbbbccccccccchhhhhhhhhhhhhhcc\n",
      "chhhhhhhhhbhbbchchhhhhhhhhhhhhhhhhcchchbbbbbccbbbbbbcbcbccbcchhhhhhhhhhhhhhhh\n",
      "METRICS -- Q3:0.753 -- MCC_h:0.639 -- MCC_b:0.712 -- MCC_c:0.589 -- MCC:0.647 -- TOT. PRED.:\n",
      "\n",
      "( 4 / 498 ) -- TOTAL ACCURACY: 0.707\n",
      "1g8q_A:\n",
      "fvnkdqiakdvkqfydqalqqavvdddannakavvktfhetldccgsstltalttsvlknnlcpsgsniisnlfkedchqkiddlfsgkh\n",
      "ccchhhhhhhhhhhhhhhhhhhhhcccchhhhhhhhhhhhhhccccccccccchhhhhhhcccccccchhhhhhccchhhhhhhhhhccc\n",
      "chchhhhhhhhhhhhhhhhhhhhhhhchhhhhhhbhbhbcbbbcccccbcbbhhhhhhbhccccccbbhhhhhhchbhhhbhbhhchbcc\n",
      "METRICS -- Q3:0.733 -- MCC_h:0.625 -- MCC_b:0.000 -- MCC_c:0.608 -- MCC:0.411 -- TOT. PRED.:\n",
      "\n",
      "( 5 / 498 ) -- TOTAL ACCURACY: 0.711\n",
      "1kdv_A:\n",
      "gtakghnptefptiydassaptaanttvgiitiggvsqtlqdlqqftsanglasvntqtiqtgssngdysddqqgqgewdldsqsivgsaggavqqllfymadqsasgntgltqafnqavsdnvakvinvslgwceadanadgtlqaedrifataaaqgqtfsvssgdegvyecnnrgypdgstysvswpasspnviavggttlyttsagaysnetvwnegldsngklwatgggysvyeskpswqsvvsgtpgrrllpdisfdaaqgtgaliynygqlqqiggtslaspifvglwarlqsansnslgfpaasfysaisstpslvhdvksgnngyggygynagtgwdyptgwgsldiaklsayirsngf\n",
      "cbbccccccchhhhhcccccccbbbbbbbbbbbccchhhhhhhhhhhhhhccccccbbbbbccccccccccchhhhhhhhhhhhhhhhhhcccbbbbbbbbbcccccchhhhhhhhhhhhhhccccbbbbcccbbhhhhhhhchhhhhhhhhhhhhhhccbbbbbccccccccccccccccccccccbbcccccccbbbbbbbbbbcccccbbbbbbcbbbbcccccbbbcccbbbbbbbcccccccccccccbbbbccbbbbccccccbbbbbccbbbbbcchhhhhhhhhhhhhhhhhhccccccchhhhhhhhhhhccccbbcccccccccccccccccccccccccccbbchhhhhhhhhhhcc\n",
      "bbbcccccccchbbbhccccccbcbhbcbbbbbbbhbhchhhhhhhhhhhchchcbbbbbbbbbcccbccbhbhbchbhbcbbchbhhhcchbhhhbhbhhhhccchcbhhhbhhhhhhhhhchbhbbbbccbhhbhhhhcchhhhhhhhhhhhhhhhcbbbbbbccccbbbbhcccccccbcbbbbbbcccchccbbbbbbbbbbbbccccbcbbbbbbbccccbbbbbbbcbbbbbbbcccbcbbcbcccccbbccccbbhhhchhhcbhbbhhbbbhccccbccchchhhhhbhhhhcchhhchcccbhbbhbhhhcchbbbbhbbcccbcbbbbbbcbcbcccccbcbccbhhhhhhhhhhccc\n",
      "METRICS -- Q3:0.628 -- MCC_h:0.517 -- MCC_b:0.429 -- MCC_c:0.425 -- MCC:0.457 -- TOT. PRED.:\n",
      "\n",
      "( 6 / 498 ) -- TOTAL ACCURACY: 0.677\n",
      "1n13_B:\n",
      "imppeaeivplpklpmgalvptaygyiisdvpgetisaaisvaipkdkslcglimeyegkcskkeaektvremakigfemrgweldriesiavehtveklgcafaaaalwyk\n",
      "ccccccbbccccccccccbbbbbbbbbbbccccbbbbbbbbbbbbcccccccbbbbbbbccchhhhhhhhhhhhhhhhhhhcccbbbbbbbbbbbbbcccbbbbbbbbbbcc\n",
      "ccbcccbcccccccccccbcbbbcbbbbbbccccchbbbhhbbbbccchchhbhhhhbhbchhhhhhhhhhhhhhhhhbhhhcbchhhbbhhhbhhhhhhhhhhhhhhhhhh\n",
      "METRICS -- Q3:0.571 -- MCC_h:0.413 -- MCC_b:0.365 -- MCC_c:0.589 -- MCC:0.455 -- TOT. PRED.:\n",
      "\n",
      "( 7 / 498 ) -- TOTAL ACCURACY: 0.665\n",
      "1tuk_A:\n",
      "acqasqlavcasailsgakpsgeccgnlraqqgcfcqyakdptygqyirsphardtltscglavphc\n",
      "cccccccccchhhhhhccccchhhhhhhhhhccchhhhhhccchhhhhhchhhhhhhhhhccccccc\n",
      "hbhhhhhhhhhhhhbhcccccchhhhccchcchhhhbccccccchcbbbchccbhchhhhbcccccc\n",
      "METRICS -- Q3:0.522 -- MCC_h:0.120 -- MCC_b:0.000 -- MCC_c:0.210 -- MCC:0.110 -- TOT. PRED.:\n",
      "\n",
      "( 8 / 498 ) -- TOTAL ACCURACY: 0.656\n",
      "3by8_A:\n",
      "dmtrdglankalavartladspeirqglqkkpqesgiqaiaeavrkrndllfivvtdmqslryshpeaqrigqpfkgddilkalngeenvainrgflaqalrvftpiydenhkqigvvaiglelsrvtqqind\n",
      "chhhhhhhhhhhhhhhhhhcchhhhhhhcccccccchhhhhhhhhhhhcccbbbbbccccbbccccccccccbbccccccccccccccbbbbbccccccbbbbbbbbbcccccbbbbbbbbbbhhhhhhhhhc\n",
      "hhhhhhhhhhhhhhhhhhccchhhhhhhbhccchhhhhhhhhhhhhchhhbbbbbcchchhbcbchhhhhhccbcchchhhhhhhccchbhbbhhhhhhhhbbbbhchbccccbhbbbbbbbbhhhhhhhhbh\n",
      "METRICS -- Q3:0.639 -- MCC_h:0.485 -- MCC_b:0.662 -- MCC_c:0.386 -- MCC:0.511 -- TOT. PRED.:\n",
      "\n",
      "( 9 / 498 ) -- TOTAL ACCURACY: 0.654\n",
      "1xg0_A:\n",
      "amdsakapqitifdhrgcsrapkestggkaggqddemmvkvastkvtvsesdaakklqefitfekgidgpftskn\n",
      "cccccbbbbbbbbbccccccccccccccccccccccbbbbbbbbbcccchhhhhhhhhhhhcccccccccccccc\n",
      "hhhhcccccbbbbbbchccccccbhchhcccccchhbbbbhbhhbbbbhhhhhhhhhhhhbbbbbcccbcccbcc\n",
      "METRICS -- Q3:0.600 -- MCC_h:0.540 -- MCC_b:0.371 -- MCC_c:0.386 -- MCC:0.432 -- TOT. PRED.:\n",
      "\n",
      "( 10 / 498 ) -- TOTAL ACCURACY: 0.651\n",
      "1c4q_A:\n",
      "tpdcvtgkveytkyndddtftvkvgdkelatnranlqslllsaqitgmtvtiktnachngggfsevifr\n",
      "ccbbbbbbbbbbbbcccccbbbbbccbbbbbccchhhhhhhhhhhhccbbbbbcccccccccccbbbbc\n",
      "cccbccbbbbbbbbccccbbbbbbccbhhhhhhhhhhhhhhhhhhccbbbbbbbcbbccccccbbbbbb\n",
      "METRICS -- Q3:0.739 -- MCC_h:0.685 -- MCC_b:0.590 -- MCC_c:0.558 -- MCC:0.611 -- TOT. PRED.:\n",
      "\n",
      "( 11 / 498 ) -- TOTAL ACCURACY: 0.656\n",
      "2fdn_A:\n",
      "ayvineaciscgacepecpvnaissgddryvidadtcidcgacagvcpvdapvqa\n",
      "cbbbccccccccccccccccccbbccccccbbccccccccchhhhhcccccbbbc\n",
      "bbbbbccccccccccccccbbbbbcccccbbbbccccccchccccccccccbhch\n",
      "METRICS -- Q3:0.709 -- MCC_h:-0.076 -- MCC_b:0.558 -- MCC_c:0.356 -- MCC:0.279 -- TOT. PRED.:\n",
      "\n",
      "( 12 / 498 ) -- TOTAL ACCURACY: 0.658\n",
      "1wt6_A:\n",
      "evtlrelqealeeevltrqslsremeairtdnqnfasqlreaearnrdleahvrqlqermellqa\n",
      "ccchhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhc\n",
      "hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh\n",
      "METRICS -- Q3:0.938 -- MCC_h:0.000 -- MCC_b:0.000 -- MCC_c:0.000 -- MCC:0.000 -- TOT. PRED.:\n",
      "\n",
      "( 13 / 498 ) -- TOTAL ACCURACY: 0.670\n",
      "1fsg_A:\n",
      "gshmaskpiedygkgkgriepmyipdntfynaddflvpphckpyidkillpgglvkdrveklaydihrtyfgeelhiicilkgsrgffnllidylatiqkysgressvppffehyvrlksyqndnstgqltvlsddlsifrdkhvlivedivdtgftltefgerlkavgpksmriatlvekrtdrsnslkgdfvgfsiedvwivgccydfnemfrdfdhvavlsdaarkkfek\n",
      "cccccccccccccccccccccbbccccccbbcccccccccccccccbbbbchhhhhhhhhhhhhhhhhhhccccbbbbbbbcchhhhhhhhhhhhhhhhhhccccccccccbbbbbbbbbbbccbbbbbbbbbbcccccccccbbbbbbbbbccchhhhhhhhhhhhccccbbbbbbbbbbcccccccccccbbbbbbccccbbcccccccccccccccbbbcchhhhhhhcc\n",
      "bbccchccchhhchbccbbbbbbccccccbbccccbcbcchccchhbbbccccchhhbhhhhhhhhhhhhchccbbbbbbbcccchhhhbhbhhhhhhhhhbcccbbcbcchbhhhhbbbhccccccbbbbbbbhchhhhhhhhbbbbbbbbbcccbhbhhhhhhhhhccchbbhbhbhhhchccchhbcbbbbbbbbbbbbbbbbhbchcchchhhhbbbhhhhhhhhhhhh\n",
      "METRICS -- Q3:0.627 -- MCC_h:0.476 -- MCC_b:0.510 -- MCC_c:0.441 -- MCC:0.476 -- TOT. PRED.:\n",
      "\n",
      "( 14 / 498 ) -- TOTAL ACCURACY: 0.664\n",
      "1utg_A:\n",
      "gicprfahvienlllgtpssyetslkefepddtmkdagmqmkkvldslpqttrenimkltekivksplcm\n",
      "ccchhhhhhhhhhhhcchhhhhhhhhhcccchhhhhhhhhhhhhhhccchhhhhhhhhhhhhhhhccccc\n",
      "ccchchchhhhhhhbccchchhhbhhhcccchhhhhhhhhhhhhhhccchchhhhhbhhhhhhhccbccc\n",
      "METRICS -- Q3:0.857 -- MCC_h:0.737 -- MCC_b:0.000 -- MCC_c:0.765 -- MCC:0.501 -- TOT. PRED.:\n",
      "\n",
      "( 15 / 498 ) -- TOTAL ACCURACY: 0.672\n",
      "3a8g_A:\n",
      "enaapaqapvsdrawalfraldgkglvpdgyvegwkktfeedfsprrgaelvarawtdpefrqllltdgtaavaqygylgpqgeyivavedtptlknvivcslatawpilglpptwyksfeyrarvvreprkvlsemgteiasdieirvydttaetrymvlpqrpagtegwsqeqlqeivtkdcligvaipqvpt\n",
      "ccccccccchhhhhhhhhhhhhhccccccchhhhhhhhhhhhcchhhhhhhhhhhhhchhhhhhhhhchhhhhhhhcccccccccbbbbbccccbbbbbbcccccchhhhccccccccchhhhhhhccchhhhhhhhccccccccbbbbbbcccccbbbbbccccccccccchhhhhhhcchhhhhccccccccc\n",
      "hhcccccchhhhhhhhhhhhhcbccbcccbcbhhcbhhbbcccchhhhhhhhbhhhcchhhhhbhhhchchhhhhhbccccccbbbbbbbccccccbbhbcbhcccccccccbcccchbhhbbbhhhchbbhhhhhhcchhhhhhbbbbbbcccchbbbbbccccccccchchhhhhhhhhhchcbbbbcbcccc\n",
      "METRICS -- Q3:0.703 -- MCC_h:0.541 -- MCC_b:0.544 -- MCC_c:0.540 -- MCC:0.542 -- TOT. PRED.:\n",
      "\n",
      "( 16 / 498 ) -- TOTAL ACCURACY: 0.675\n",
      "2pof_A:\n",
      "sdtlrkivleeclpnqqqnqnpspcaevkpnagyvvlkdlngplqyllptyringtesplltdpstpnffwlawqardfskkygqpvpdravslainsrtgrtqnhfhihiscirpdvreqldnnlanissrwlplpgglrgheylarrvteselvqrspflaeevpearehgsyglavrqsdnsfvllatqrnlltlnrasaeeiqdhqceil\n",
      "chhhhhhhhhhhhhhhhhhcccccbbbbbccccbbbbbcccccccbbbcccccccccccccccccccchhhhhhhhcccccccccccccccbbbbcccccccccccccbbbbbcchhhhhhhhhcccccccccbbcccccccccbbbbbbchhhhhhcccccccccccccccccbbbbbbcccccbbbbbbbcccccccccccccccccccccc\n",
      "chhhhhbhbhhhchhhhhhccccccbbcbccchhbbbbchcchcbcbbcbbbccccccccbccccchchhhbhhhhchhbhccccccchchhbbbhbhccbccbchcbbbbbbcchchhhbhhhchhhcbbcccbcccccchbhbhhhhhhhhhhchhchchhhchhhhhhcchhbbbbbchchhhhbhbhhhchhhhhhhhhhhhhhhchhhh\n",
      "METRICS -- Q3:0.579 -- MCC_h:0.353 -- MCC_b:0.505 -- MCC_c:0.424 -- MCC:0.428 -- TOT. PRED.:\n",
      "\n",
      "( 17 / 498 ) -- TOTAL ACCURACY: 0.665\n",
      "2cc6_A:\n",
      "vfkkvlltgtseesftaaaddaidraedtldnvvwaevvdqgveigaveertyqtevqvafeld\n",
      "cbbbbbbbbbbccchhhhhhhhhhhhhhccccbbbbbbbbbbbbcccccccbbbbbbbbbbbcc\n",
      "hbbbbbbbcccchhhhhhhhhhhhhhbhhhhhbbbbbbcbcbbbhhbhbhhhbbhchhhhbbhb\n",
      "METRICS -- Q3:0.547 -- MCC_h:0.454 -- MCC_b:0.520 -- MCC_c:-0.097 -- MCC:0.292 -- TOT. PRED.:\n",
      "\n",
      "( 18 / 498 ) -- TOTAL ACCURACY: 0.662\n",
      "2ohw_A:\n",
      "edkmdlylqqgmygpletkpderhlflgslrervvlaltkgqvlrskpykeaehelknshnvtllingelqyqsyssyiqmasrygvpfkivsdlqfhtplgiviaadiavnreliyiqddiynrsvl\n",
      "chhhhhhhhhcccccccchhhhhhhhcccccccbbbbbbhhhhhccccchhhhhhhhhcccbbbbbbccccccchhhhhhhhhhhcccbbbbccccccccbbbbbbbccccccccccccchhhhhhhc\n",
      "hhcbhhhhhcchhhccccchchhhhhhhhhhhhhhbhhhhhhhbchchhhhhhhbhhhhcbbbbbhhhccbhhchhhhhhhhhbhccbbbbbbbcccbcccbbbhhhchhhhhhhbbbbhhhbhhhhb\n",
      "METRICS -- Q3:0.555 -- MCC_h:0.357 -- MCC_b:0.343 -- MCC_c:0.343 -- MCC:0.348 -- TOT. PRED.:\n",
      "\n",
      "( 19 / 498 ) -- TOTAL ACCURACY: 0.656\n",
      "2bog_X:\n",
      "ndspfyvnpnmssaewvrnnpndprtpvirdriasvpqgtwfahhnpgqitgqvdalmsaaqaagkipilvvsnapgrdcgapshsayrswidefaaglknrpayiivepdlislmsscmqhvqqevletmayagkalkagssqariyfdaghsawhspaqmaswlqqadisnsahgiatntsnyrwtadevayakavlsaignpslravidtsrngngpagnewcdpsgraigtpsttntgdpmidaflwiklpgeadgciagagqfvpqaayemaiaa\n",
      "ccccccccccchhhhhhhhccccchhhhhhhhhccccccbbbcccccchhhhhhhhhhhhhhhhccccbbbbcccccccccccchhhhhhhhhhhhhhcccccbbbbbcccccccccccchhhhhhhhhhhhhhhhhhhhhcccbbbbbbcccccccchhhhhhhhhhhcccccccbbbbccccccchhhhhhhhhhhhhhhcccccbbbbbcccccccccccccccccccccccccccccccccbbbbbccccccbbccccccccbbchhhhhhhhhhc\n",
      "ccccbbbbccbccchhhbhccccccchbhbhchcccccbbbbbbcccbbcbhbhhhhhhhhhhhccccbbbbbccccccccccccbbhhhhhhhhhhhhcccccbbbbcccchhhbhhhhhhhhhhhhhhhhhhhhhhhhcchchbbbbbcccccccchchhhhhhhhhchbhhccbbbbcchbccchhhhhhhhhhhbhbhcchcbbbhbbhbccccccccccccchcccbcbcbbcbbccchhhhbhbbbcccccchhbbchccccchhhhhhhhhhh\n",
      "METRICS -- Q3:0.725 -- MCC_h:0.646 -- MCC_b:0.469 -- MCC_c:0.572 -- MCC:0.563 -- TOT. PRED.:\n",
      "\n",
      "( 20 / 498 ) -- TOTAL ACCURACY: 0.664\n",
      "256b_A:\n",
      "adlednmetlndnlkviekadnaaqvkdaltkmraaaldaqkatppkledkspdspemkdfrhgfdilvgqiddalklanegkvkeaqaaaeqlkttrnayhqkyr\n",
      "cchhhhhhhhhhhhhhhhhccchhhhhhhhhhhhhhhhhhhcccccccccccccchhhhhhhhhhhhhhhhhhhhhhhhhhcchhhhhhhhhhhhhhhhhhhhhhc\n",
      "hhhhhchhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhccbcccccccccchchchhbhhhbbbhhhhhhhhhhhhhhhhhhhhhhhhhhhhhchhhhhbhc\n",
      "METRICS -- Q3:0.830 -- MCC_h:0.521 -- MCC_b:0.000 -- MCC_c:0.600 -- MCC:0.374 -- TOT. PRED.:\n",
      "\n",
      "( 21 / 498 ) -- TOTAL ACCURACY: 0.670\n",
      "2ggc_A:\n",
      "aisiktpediekmrvagrlaaevlemiepyvkpgvstgeldricndyivneqhavsaclgyhgypksvcisinevvchgipddakllkdgdivnidvtvikdgfhgdtskmfivgkptimgerlcritqeslylalrmvkpginlreigaaiqkfveaegfsvvreycghgigrgfheepqvlhydsretnvvlkpgmtftiepmvnagkkeirtmkdgwtvktkdrslsaqyehtivvtdngceiltlrkddtipaiishde\n",
      "cccccchhhhhhhhhhhhhhhhhhhhhcccccccbbhhhhhhhhhhhhhhhhccbbcccccccccccbbbbbccbbbccccccccbbcccbbbbbbbbbbbccbbbbbbbbbbbccccccchhhhhhhhhhhhhhhhhccccbbhhhhhhhhhhhhhhhccbbcccccbbbcccccccccbbccccccccccbbcccbbbbbccbbbcccccbbbccccccbbbcccccbbbccbbbbbbccbbbbcccccccccccbbbccc\n",
      "bbbbcchhhhhhhhhhhhhhhhhhhhhcccbcccccbcbbhhhhchhhcccbhhhhbhbhccccccbbbbbhbbbbbccccccchbccccbbbbbbbbbbbcbbcbccbbbbbbbcccccchhhbbchhhhhhhhhhhhccccbhhhhhhhhhhhhhhcccbbhbbchbhbcbcbcbcccbbbhbbccbhhbbbbccbbbbbbhhhhcchbbhbbhbhccbbbcbchchhhhhcbbbbbbhccccbbbbbccccccbbhbccc\n",
      "METRICS -- Q3:0.662 -- MCC_h:0.609 -- MCC_b:0.462 -- MCC_c:0.455 -- MCC:0.509 -- TOT. PRED.:\n",
      "\n",
      "( 22 / 498 ) -- TOTAL ACCURACY: 0.669\n",
      "3rnq_A:\n",
      "sltfypawltvseganatftcslsnwsedlmlnwnrlspsnqtekqaafsnglsqpvqdarfqiiqlpnrhdfhmnildtrrndsgiylcgaisrhpkakieespgaelvvt\n",
      "cbbbbccbbbbcccccbbbbbbbccccccbbbbbbbbcccccbbbbbbbbccbbbbcccccbbbbbcccccbbbbbbccccccccbbbbbbbbbcccccccbbcccbbbbbc\n",
      "cbccccchbbhccccccbbbbbbccccchhhbbhhcccccchhhhhhhhhhccbchhchhhbbhbbccccchbbbbbhbhccccchbbbbbbhbccbccchbcchccbbbbb\n",
      "METRICS -- Q3:0.616 -- MCC_h:0.000 -- MCC_b:0.503 -- MCC_c:0.565 -- MCC:0.356 -- TOT. PRED.:\n",
      "\n",
      "( 23 / 498 ) -- TOTAL ACCURACY: 0.667\n",
      "2iyv_A:\n",
      "apkavlvglpgsgkstigrrlakalgvglldtdvaieqrtgrsiadifatdgeqefrrieedvvraaladhdgvlslgggavtspgvraalaghtvvyleisaaegvrrtggntvrpllagpdraekyralmakraplyrrvatmrvdtnrrnpgavvrhilsrlqvpspseaatlehh\n",
      "cccbbbbcccccchhhhhhhhhhhhcccbbbhhhhhhhhhccchhhhhhhhchhhhhhhhhhhhhhhhhhcccbbbcccccccchhhhhhhccccbbbbbcchhhhhhhcccccccccccccchhhhhhhhhhhhhhhhhhhccbbbbccccchhhhhhhhhhccccccchhhhhhhcc\n",
      "cbhbbbbbcccccchchhhhhhhhhccbcbbhbhhhhhhhhchhbhbhhhchhhhhhhhbhhhhhhhhhchchbbhbbcccbcchhhhhhhchchbbbbbhhhhhcchhbcccbbbbhbcchchhhhhhhhhhhhchhhhhhbbbbbbbccccccbbhbbhhbbccccchcchhhhhhh\n",
      "METRICS -- Q3:0.704 -- MCC_h:0.597 -- MCC_b:0.537 -- MCC_c:0.449 -- MCC:0.527 -- TOT. PRED.:\n",
      "\n",
      "( 24 / 498 ) -- TOTAL ACCURACY: 0.669\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-36-117b14a5c73a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[1;31m# Make the predictions.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0mprediction_list\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpredict_sec_structures\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0minput_data\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdata_type\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m \u001b[1;31m# Overall Q3\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0mq3_tot\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmean\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mq3\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mprediction_list\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-34-59a8e50fed09>\u001b[0m in \u001b[0;36mpredict_sec_structures\u001b[0;34m(input_data, data_type, print_details)\u001b[0m\n\u001b[1;32m     37\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ma_list\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m             \u001b[1;31m# Predict the secondary structure for a given aminoacid of the protein.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m             \u001b[0mres\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcompute_gor_3\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ma_list\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mi\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msj_aj_ajm_dict\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msj_aj_matrix\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0ma_occ\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0ml_1_out\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m             \u001b[1;31m# Add the new secondary structure that is predicted\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m             \u001b[0mpred\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[0mres\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-32-8c1aeaabcad8>\u001b[0m in \u001b[0;36mcompute_gor_3\u001b[0;34m(a_list, i, sj_aj_ajm_dict, sj_aj_matrix, a_occ, l_1_out)\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m             \u001b[1;31m# Number of times aminoacid a_j appears together with structure s_j.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m             \u001b[0mden_2\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0msj_aj_matrix\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mat\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0ms\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0ma_j\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m-\u001b[0m \u001b[0msub_2\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m             \u001b[1;31m# Number of times aminoacid a_j appears with a different structure from s_j.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m             \u001b[0mnum_2\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0ma_occ\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mat\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0ma_j\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m-\u001b[0m \u001b[0mden_2\u001b[0m \u001b[1;33m-\u001b[0m \u001b[0msub_2\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\albyr\\Anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1663\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m   1664\u001b[0m         \u001b[0mkey\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_convert_key\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1665\u001b[0;31m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_value\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtakeable\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_takeable\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1666\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m   1667\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m__setitem__\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mC:\\Users\\albyr\\Anaconda3\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36mget_value\u001b[0;34m(self, index, col, takeable)\u001b[0m\n\u001b[1;32m   1897\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m   1898\u001b[0m         \u001b[0mseries\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_item_cache\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1899\u001b[0;31m         \u001b[0mengine\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1900\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mengine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_value\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mseries\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_values\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m   1901\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "    # Type of the data to read (\"stride\", \"dssp\")\n",
    "    data_type = \"stride\"\n",
    "    # File name\n",
    "    file_name = \"../data/\" + data_type + \"_info.txt\"\n",
    "    # Read the data\n",
    "    input_data = preprocess_input(file_name, aminoacid_codes)\n",
    "\n",
    "    # Append the PDB chain code to the PDB code, to obtain unique protein identifier.\n",
    "    input_data.PDB_code = input_data.PDB_code + \"_\" + input_data.PDB_chain_code\n",
    "\n",
    "    # Make the predictions.\n",
    "    # UNCOMMENT IF NECESSARY\n",
    "    # prediction_list = predict_sec_structures(input_data, data_type, True)\n",
    "    \n",
    "    # Overall Q3\n",
    "    q3_tot = np.mean([x.q3 for x in prediction_list])\n",
    "    print(\"\\nOVERALL Q3:\", q3_tot)\n",
    "\n",
    "    # Load CATH file\n",
    "    cath = pd.read_csv(\"../data/cath_info.txt\", header=None, sep=\"\\t\",\n",
    "                         names=[\"PDB_code\", \"PDB_chain\", \"overall_structure\"])\n",
    "     \n",
    "    # Add a composite key to the cath data\n",
    "    cath[\"PDB_code_and_chain\"] = cath.PDB_code + \"_\" + cath.PDB_chain\n",
    "    \n",
    "    # Save predictions in a file.\n",
    "    result_frame = pd.DataFrame(index=range(len(prediction_list)), columns=[\"PDB_code_and_chain\",\n",
    "                                                                            \"length\",\n",
    "                                                                            \"prediction\",\n",
    "                                                                            \"real_sec_structure\",\n",
    "                                                                            \"num_predicted_h\",\n",
    "                                                                            \"num_predicted_b\",\n",
    "                                                                            \"num_predicted_c\",\n",
    "                                                                            \"num_real_h\",\n",
    "                                                                            \"num_real_b\",\n",
    "                                                                            \"num_real_c\",\n",
    "                                                                            \"num_am_a\",\n",
    "                                                                            \"num_am_r\",\n",
    "                                                                            \"num_am_n\",\n",
    "                                                                            \"num_am_d\",\n",
    "                                                                            \"num_am_c\",\n",
    "                                                                            \"num_am_q\",\n",
    "                                                                            \"num_am_e\",\n",
    "                                                                            \"num_am_g\",\n",
    "                                                                            \"num_am_h\",\n",
    "                                                                            \"num_am_i\",\n",
    "                                                                            \"num_am_l\",\n",
    "                                                                            \"num_am_k\",\n",
    "                                                                            \"num_am_m\",\n",
    "                                                                            \"num_am_f\",\n",
    "                                                                            \"num_am_p\",\n",
    "                                                                            \"num_am_s\",\n",
    "                                                                            \"num_am_t\",\n",
    "                                                                            \"num_am_w\",\n",
    "                                                                            \"num_am_y\",\n",
    "                                                                            \"num_am_v\",\n",
    "                                                                            \"protein\",\n",
    "                                                                            \"q3\",\n",
    "                                                                            \"mcc_h\",\n",
    "                                                                            \"mcc_b\",\n",
    "                                                                            \"mcc_c\",\n",
    "                                                                            \"mcc\",\n",
    "                                                                            \"overall_pred\",\n",
    "                                                                            \"overall_structure_real\"])\n",
    "    for p_i, p in enumerate(prediction_list):\n",
    "        result_frame.iloc[p_i, 0] = p.pdb_code\n",
    "        result_frame.iloc[p_i, 1] = len(p.prediction)\n",
    "        result_frame.iloc[p_i, 2] = p.prediction\n",
    "        result_frame.iloc[p_i, 3] = p.real_sec_structure\n",
    "        result_frame.iloc[p_i, 4] = p.prediction.count(\"h\")\n",
    "        result_frame.iloc[p_i, 5] = p.prediction.count(\"b\")\n",
    "        result_frame.iloc[p_i, 6] = p.prediction.count(\"c\")\n",
    "        result_frame.iloc[p_i, 7] = p.real_sec_structure.count(\"h\")\n",
    "        result_frame.iloc[p_i, 8] = p.real_sec_structure.count(\"b\")\n",
    "        result_frame.iloc[p_i, 9] = p.real_sec_structure.count(\"c\")\n",
    "        result_frame.iloc[p_i, 10] = p.protein.count(\"a\")\n",
    "        result_frame.iloc[p_i, 11] = p.protein.count(\"r\")\n",
    "        result_frame.iloc[p_i, 12] = p.protein.count(\"n\")\n",
    "        result_frame.iloc[p_i, 13] = p.protein.count(\"d\")\n",
    "        result_frame.iloc[p_i, 14] = p.protein.count(\"c\")\n",
    "        result_frame.iloc[p_i, 15] = p.protein.count(\"q\")\n",
    "        result_frame.iloc[p_i, 16] = p.protein.count(\"e\")\n",
    "        result_frame.iloc[p_i, 17] = p.protein.count(\"g\")\n",
    "        result_frame.iloc[p_i, 18] = p.protein.count(\"h\")\n",
    "        result_frame.iloc[p_i, 19] = p.protein.count(\"i\")\n",
    "        result_frame.iloc[p_i, 20] = p.protein.count(\"l\")\n",
    "        result_frame.iloc[p_i, 21] = p.protein.count(\"k\")\n",
    "        result_frame.iloc[p_i, 22] = p.protein.count(\"m\")\n",
    "        result_frame.iloc[p_i, 23] = p.protein.count(\"f\")\n",
    "        result_frame.iloc[p_i, 24] = p.protein.count(\"p\")\n",
    "        result_frame.iloc[p_i, 25] = p.protein.count(\"s\")\n",
    "        result_frame.iloc[p_i, 26] = p.protein.count(\"t\")\n",
    "        result_frame.iloc[p_i, 27] = p.protein.count(\"w\")\n",
    "        result_frame.iloc[p_i, 28] = p.protein.count(\"y\")\n",
    "        result_frame.iloc[p_i, 29] = p.protein.count(\"v\")\n",
    "        result_frame.iloc[p_i, 30] = p.protein\n",
    "        result_frame.iloc[p_i, 31] = p.q3\n",
    "        result_frame.iloc[p_i, 32] = p.mcc_h\n",
    "        result_frame.iloc[p_i, 33] = p.mcc_b\n",
    "        result_frame.iloc[p_i, 34] = p.mcc_c\n",
    "        result_frame.iloc[p_i, 35] = p.mcc\n",
    "        result_frame.iloc[p_i, 36] = p.overall_prediction\n",
    "        result_frame.iloc[p_i, 37] = cath.loc[cath['PDB_code_and_chain'] == p.pdb_code].overall_structure.iloc[0]\n",
    "    \n",
    "    # SAVE THE RESULTS TO CSV\n",
    "    # UNCOMMENT IF NECESSARY\n",
    "    # result_frame.to_csv(\"../data/pred_result_\" + data_type + \".csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
